// @credo-ts/askar/build/wallet/didcommV1.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.didcommV1Pack = didcommV1Pack;
exports.didcommV1Unpack = didcommV1Unpack;
const core_1 = require("@credo-ts/core");
const importAskar_1 = require("../utils/importAskar");
const JweEnvelope_1 = require("./JweEnvelope");
function didcommV1Pack(askarLibrary, payload, recipientKeys, senderKey) {
    let cek;
    let senderExchangeKey;
    try {
        cek = (0, importAskar_1.isOwfAskarLibrary)(askarLibrary)
            ? askarLibrary.Key.generate(askarLibrary.KeyAlgorithm.Chacha20C20P)
            : askarLibrary.Key.generate(askarLibrary.KeyAlgs.Chacha20C20P);
        const keyAlgorithm = (0, importAskar_1.isOwfAskarLibrary)(askarLibrary)
            ? askarLibrary.KeyAlgorithm.X25519
            : askarLibrary.KeyAlgs.X25519;
        senderExchangeKey = senderKey
            ? (0, importAskar_1.isOwfAskarKey)(askarLibrary, senderKey)
                ? senderKey.convertkey({ algorithm: keyAlgorithm })
                : senderKey.convertkey({ algorithm: keyAlgorithm })
            : undefined;
        const recipients = [];
        for (const recipientKey of recipientKeys) {
            let targetExchangeKey;
            try {
                targetExchangeKey = (0, importAskar_1.isOwfAskarLibrary)(askarLibrary)
                    ? askarLibrary.Key.fromPublicBytes({
                        publicKey: core_1.Key.fromPublicKeyBase58(recipientKey, core_1.KeyType.Ed25519).publicKey,
                        algorithm: askarLibrary.KeyAlgorithm.Ed25519,
                    }).convertkey({ algorithm: askarLibrary.KeyAlgorithm.X25519 })
                    : askarLibrary.Key.fromPublicBytes({
                        publicKey: core_1.Key.fromPublicKeyBase58(recipientKey, core_1.KeyType.Ed25519).publicKey,
                        algorithm: askarLibrary.KeyAlgs.Ed25519,
                    }).convertkey({ algorithm: askarLibrary.KeyAlgs.X25519 });
                if (senderKey && senderExchangeKey) {
                    const message = core_1.TypedArrayEncoder.fromString(core_1.TypedArrayEncoder.toBase58(senderKey.publicBytes));
                    const encryptedSender = (0, importAskar_1.isOwfAskarLibrary)(askarLibrary)
                        ? askarLibrary.CryptoBox.seal({
                            recipientKey: targetExchangeKey,
                            message,
                        })
                        : askarLibrary.CryptoBox.seal({
                            recipientKey: targetExchangeKey,
                            message,
                        });
                    const nonce = askarLibrary.CryptoBox.randomNonce();
                    const encryptedCek = (0, importAskar_1.isOwfAskarLibrary)(askarLibrary)
                        ? askarLibrary.CryptoBox.cryptoBox({
                            recipientKey: targetExchangeKey,
                            senderKey: senderExchangeKey,
                            message: cek.secretBytes,
                            nonce,
                        })
                        : askarLibrary.CryptoBox.cryptoBox({
                            recipientKey: targetExchangeKey,
                            senderKey: senderExchangeKey,
                            message: cek.secretBytes,
                            nonce,
                        });
                    recipients.push(new JweEnvelope_1.JweRecipient({
                        encryptedKey: encryptedCek,
                        header: {
                            kid: recipientKey,
                            sender: core_1.TypedArrayEncoder.toBase64URL(encryptedSender),
                            iv: core_1.TypedArrayEncoder.toBase64URL(nonce),
                        },
                    }));
                }
                else {
                    const encryptedCek = (0, importAskar_1.isOwfAskarLibrary)(askarLibrary)
                        ? askarLibrary.CryptoBox.seal({
                            recipientKey: targetExchangeKey,
                            message: cek.secretBytes,
                        })
                        : askarLibrary.CryptoBox.seal({
                            recipientKey: targetExchangeKey,
                            message: cek.secretBytes,
                        });
                    recipients.push(new JweEnvelope_1.JweRecipient({
                        encryptedKey: encryptedCek,
                        header: {
                            kid: recipientKey,
                        },
                    }));
                }
            }
            finally {
                targetExchangeKey?.handle.free();
            }
        }
        const protectedJson = {
            enc: 'xchacha20poly1305_ietf',
            typ: 'JWM/1.0',
            alg: senderKey ? 'Authcrypt' : 'Anoncrypt',
            recipients: recipients.map((item) => core_1.JsonTransformer.toJSON(item)),
        };
        const { ciphertext, tag, nonce } = cek.aeadEncrypt({
            message: core_1.Buffer.from(JSON.stringify(payload)),
            aad: core_1.Buffer.from(core_1.JsonEncoder.toBase64URL(protectedJson)),
        }).parts;
        const envelope = new JweEnvelope_1.JweEnvelope({
            ciphertext: core_1.TypedArrayEncoder.toBase64URL(ciphertext),
            iv: core_1.TypedArrayEncoder.toBase64URL(nonce),
            protected: core_1.JsonEncoder.toBase64URL(protectedJson),
            tag: core_1.TypedArrayEncoder.toBase64URL(tag),
        }).toJson();
        return envelope;
    }
    finally {
        cek?.handle.free();
        senderExchangeKey?.handle.free();
    }
}
function didcommV1Unpack(askarLibrary, messagePackage, recipientKey) {
    const protectedJson = core_1.JsonEncoder.fromBase64(messagePackage.protected);
    const alg = protectedJson.alg;
    if (!['Anoncrypt', 'Authcrypt'].includes(alg)) {
        throw new core_1.WalletError(`Unsupported pack algorithm: ${alg}`);
    }
    const recipient = protectedJson.recipients.find(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (r) => r.header.kid === core_1.TypedArrayEncoder.toBase58(recipientKey.publicBytes));
    if (!recipient) {
        throw new core_1.WalletError('No corresponding recipient key found');
    }
    const sender = recipient?.header.sender ? core_1.TypedArrayEncoder.fromBase64(recipient.header.sender) : undefined;
    const iv = recipient?.header.iv ? core_1.TypedArrayEncoder.fromBase64(recipient.header.iv) : undefined;
    const encrypted_key = core_1.TypedArrayEncoder.fromBase64(recipient.encrypted_key);
    if (sender && !iv) {
        throw new core_1.WalletError('Missing IV');
    }
    else if (!sender && iv) {
        throw new core_1.WalletError('Unexpected IV');
    }
    let payloadKey, senderKey;
    let sender_x;
    let recip_x;
    try {
        const keyAlgorithm = (0, importAskar_1.isOwfAskarLibrary)(askarLibrary)
            ? askarLibrary.KeyAlgorithm.X25519
            : askarLibrary.KeyAlgs.X25519;
        recip_x = (0, importAskar_1.isOwfAskarKey)(askarLibrary, recipientKey)
            ? recipientKey.convertkey({ algorithm: keyAlgorithm })
            : recipientKey.convertkey({ algorithm: keyAlgorithm });
        if (sender && iv) {
            senderKey = core_1.TypedArrayEncoder.toUtf8String((0, importAskar_1.isOwfAskarLibrary)(askarLibrary)
                ? askarLibrary.CryptoBox.sealOpen({
                    recipientKey: recip_x,
                    ciphertext: sender,
                })
                : askarLibrary.CryptoBox.sealOpen({
                    recipientKey: recip_x,
                    ciphertext: sender,
                }));
            sender_x = (0, importAskar_1.isOwfAskarLibrary)(askarLibrary)
                ? askarLibrary.Key.fromPublicBytes({
                    algorithm: askarLibrary.KeyAlgorithm.Ed25519,
                    publicKey: core_1.TypedArrayEncoder.fromBase58(senderKey),
                }).convertkey({ algorithm: askarLibrary.KeyAlgorithm.X25519 })
                : askarLibrary.Key.fromPublicBytes({
                    algorithm: askarLibrary.KeyAlgs.Ed25519,
                    publicKey: core_1.TypedArrayEncoder.fromBase58(senderKey),
                }).convertkey({ algorithm: askarLibrary.KeyAlgs.X25519 });
            payloadKey = (0, importAskar_1.isOwfAskarLibrary)(askarLibrary)
                ? askarLibrary.CryptoBox.open({
                    recipientKey: recip_x,
                    senderKey: sender_x,
                    message: encrypted_key,
                    nonce: iv,
                })
                : askarLibrary.CryptoBox.open({
                    recipientKey: recip_x,
                    senderKey: sender_x,
                    message: encrypted_key,
                    nonce: iv,
                });
        }
        else {
            payloadKey = (0, importAskar_1.isOwfAskarLibrary)(askarLibrary)
                ? askarLibrary.CryptoBox.sealOpen({ ciphertext: encrypted_key, recipientKey: recip_x })
                : askarLibrary.CryptoBox.sealOpen({ ciphertext: encrypted_key, recipientKey: recip_x });
        }
    }
    finally {
        sender_x?.handle.free();
        recip_x?.handle.free();
    }
    if (!senderKey && alg === 'Authcrypt') {
        throw new core_1.WalletError('Sender public key not provided for Authcrypt');
    }
    let cek;
    try {
        cek = (0, importAskar_1.isOwfAskarLibrary)(askarLibrary)
            ? askarLibrary.Key.fromSecretBytes({ algorithm: askarLibrary.KeyAlgorithm.Chacha20C20P, secretKey: payloadKey })
            : askarLibrary.Key.fromSecretBytes({ algorithm: askarLibrary.KeyAlgs.Chacha20C20P, secretKey: payloadKey });
        const message = cek.aeadDecrypt({
            ciphertext: core_1.TypedArrayEncoder.fromBase64(messagePackage.ciphertext),
            nonce: core_1.TypedArrayEncoder.fromBase64(messagePackage.iv),
            tag: core_1.TypedArrayEncoder.fromBase64(messagePackage.tag),
            aad: core_1.TypedArrayEncoder.fromString(messagePackage.protected),
        });
        return {
            plaintextMessage: core_1.JsonEncoder.fromBuffer(message),
            senderKey,
            recipientKey: core_1.TypedArrayEncoder.toBase58(recipientKey.publicBytes),
        };
    }
    finally {
        cek?.handle.free();
    }
}
