// @hyperledger/indy-vdr-nodejs/build/ffi/primitives.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FFI_POOL_HANDLE = exports.FFI_REQUEST_HANDLE = exports.FFI_HANDLE = exports.FFI_ERROR_CODE = exports.FFI_CALLBACK_ID = exports.FFI_VOID = exports.FFI_STRING = exports.FFI_INT64 = exports.FFI_INT32 = exports.FFI_INT8 = exports.FFI_USIZE = exports.FFI_UINT64 = exports.FFI_UINT8 = void 0;
const koffi = __importStar(require("koffi"));
exports.FFI_UINT8 = koffi.types.uint8_t;
exports.FFI_UINT64 = koffi.types.uint64_t;
exports.FFI_USIZE = koffi.types.size_t;
exports.FFI_INT8 = koffi.types.int8_t;
exports.FFI_INT32 = koffi.types.int32_t;
exports.FFI_INT64 = koffi.types.int64_t;
exports.FFI_STRING = koffi.types.string;
exports.FFI_VOID = koffi.types.void;
exports.FFI_CALLBACK_ID = exports.FFI_INT64;
exports.FFI_ERROR_CODE = exports.FFI_INT64;
exports.FFI_HANDLE = exports.FFI_INT64;
exports.FFI_REQUEST_HANDLE = exports.FFI_HANDLE;
exports.FFI_POOL_HANDLE = exports.FFI_HANDLE;
