// @credo-ts/core/build/modules/oob/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertToNewInvitation = convertToNewInvitation;
exports.convertToOldInvitation = convertToOldInvitation;
const connections_1 = require("../connections");
const helpers_1 = require("../dids/helpers");
const OutOfBandDidCommService_1 = require("./domain/OutOfBandDidCommService");
const messages_1 = require("./messages");
function convertToNewInvitation(oldInvitation) {
    let service;
    if (oldInvitation.did) {
        service = oldInvitation.did;
    }
    else if (oldInvitation.serviceEndpoint && oldInvitation.recipientKeys && oldInvitation.recipientKeys.length > 0) {
        service = new OutOfBandDidCommService_1.OutOfBandDidCommService({
            id: '#inline',
            recipientKeys: oldInvitation.recipientKeys?.map(helpers_1.verkeyToDidKey),
            routingKeys: oldInvitation.routingKeys?.map(helpers_1.verkeyToDidKey),
            serviceEndpoint: oldInvitation.serviceEndpoint,
        });
    }
    else {
        throw new Error('Missing required serviceEndpoint, routingKeys and/or did fields in connection invitation');
    }
    const options = {
        id: oldInvitation.id,
        label: oldInvitation.label,
        imageUrl: oldInvitation.imageUrl,
        appendedAttachments: oldInvitation.appendedAttachments,
        accept: ['didcomm/aip1', 'didcomm/aip2;env=rfc19'],
        services: [service],
        // NOTE: we hardcode it to 1.0, we won't see support for newer versions of the protocol
        // and we also can process 1.0 if we support newer versions
        handshakeProtocols: ['https://didcomm.org/connections/1.0'],
    };
    const outOfBandInvitation = new messages_1.OutOfBandInvitation(options);
    outOfBandInvitation.invitationType = messages_1.InvitationType.Connection;
    return outOfBandInvitation;
}
function convertToOldInvitation(newInvitation) {
    // Taking first service, as we can only include one service in a legacy invitation.
    const [service] = newInvitation.getServices();
    let options;
    if (typeof service === 'string') {
        options = {
            id: newInvitation.id,
            // label is optional
            label: newInvitation.label ?? '',
            did: service,
            imageUrl: newInvitation.imageUrl,
            appendedAttachments: newInvitation.appendedAttachments,
        };
    }
    else {
        options = {
            id: newInvitation.id,
            // label is optional
            label: newInvitation.label ?? '',
            recipientKeys: service.recipientKeys.map(helpers_1.didKeyToVerkey),
            routingKeys: service.routingKeys?.map(helpers_1.didKeyToVerkey),
            serviceEndpoint: service.serviceEndpoint,
            imageUrl: newInvitation.imageUrl,
            appendedAttachments: newInvitation.appendedAttachments,
        };
    }
    const connectionInvitationMessage = new connections_1.ConnectionInvitationMessage(options);
    return connectionInvitationMessage;
}
