// @hyperledger/aries-askar-shared/build/store/EntryList.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntryList = void 0;
const ariesAskar_1 = require("../ariesAskar");
const Entry_1 = require("./Entry");
class EntryList {
    constructor({ handle, length }) {
        this._length = 0;
        this._handle = handle;
        this._length = length || ariesAskar_1.ariesAskar.entryListCount({ entryListHandle: handle });
    }
    get handle() {
        return this._handle;
    }
    get length() {
        return this._length;
    }
    getEntryByIndex(index) {
        return new Entry_1.Entry({ list: this.handle, position: index });
    }
    forEach(cb) {
        for (let i = 0; i < this.length; i++) {
            cb(this.getEntryByIndex(i), i);
        }
    }
    find(cb) {
        for (let i = 0; i < this.length; i++) {
            const item = this.getEntryByIndex(i);
            if (cb(item))
                return item;
        }
    }
    toArray(valuesAreJson) {
        const list = [];
        this.forEach((entry) => list.push(entry.toJson(valuesAreJson)));
        return list;
    }
}
exports.EntryList = EntryList;
