// @hyperledger/anoncreds-shared/build/api/CredentialDefinition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CredentialDefinition = void 0;
const AnoncredsObject_1 = require("../AnoncredsObject");
const register_1 = require("../register");
const CredentialDefinitionPrivate_1 = require("./CredentialDefinitionPrivate");
const KeyCorrectnessProof_1 = require("./KeyCorrectnessProof");
const Schema_1 = require("./Schema");
const utils_1 = require("./utils");
class CredentialDefinition extends AnoncredsObject_1.AnoncredsObject {
    static create(options) {
        var _a;
        let createReturnObj;
        // Objects created within this method must be freed up
        const objectHandles = [];
        try {
            const schema = options.schema instanceof Schema_1.Schema
                ? options.schema.handle
                : (0, utils_1.pushToArray)(Schema_1.Schema.fromJson(options.schema).handle, objectHandles);
            createReturnObj = register_1.NativeAnoncreds.instance.createCredentialDefinition({
                schemaId: options.schemaId,
                schema,
                signatureType: options.signatureType,
                tag: options.tag,
                issuerId: options.issuerId,
                supportRevocation: (_a = options.supportRevocation) !== null && _a !== void 0 ? _a : false,
            });
        }
        finally {
            for (const handle of objectHandles) {
                handle.clear();
            }
        }
        return {
            credentialDefinition: new CredentialDefinition(createReturnObj.credentialDefinition.handle),
            credentialDefinitionPrivate: new CredentialDefinitionPrivate_1.CredentialDefinitionPrivate(createReturnObj.credentialDefinitionPrivate.handle),
            keyCorrectnessProof: new KeyCorrectnessProof_1.KeyCorrectnessProof(createReturnObj.keyCorrectnessProof.handle),
        };
    }
    static fromJson(json) {
        return new CredentialDefinition(register_1.NativeAnoncreds.instance.credentialDefinitionFromJson({ json: JSON.stringify(json) }).handle);
    }
}
exports.CredentialDefinition = CredentialDefinition;
