// @credo-ts/core/build/modules/basic-messages/repository/BasicMessageRecord.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BasicMessageRecord = void 0;
const BaseRecord_1 = require("../../../storage/BaseRecord");
const uuid_1 = require("../../../utils/uuid");
class BasicMessageRecord extends BaseRecord_1.BaseRecord {
    constructor(props) {
        super();
        this.type = BasicMessageRecord.type;
        if (props) {
            this.id = props.id ?? (0, uuid_1.uuid)();
            this.createdAt = props.createdAt ?? new Date();
            this.content = props.content;
            this.sentTime = props.sentTime;
            this.connectionId = props.connectionId;
            this._tags = props.tags ?? {};
            this.role = props.role;
            this.threadId = props.threadId;
            this.parentThreadId = props.parentThreadId;
        }
    }
    getTags() {
        return {
            ...this._tags,
            connectionId: this.connectionId,
            role: this.role,
            threadId: this.threadId,
            parentThreadId: this.parentThreadId,
        };
    }
}
exports.BasicMessageRecord = BasicMessageRecord;
BasicMessageRecord.type = 'BasicMessageRecord';
