// @hyperledger/aries-askar-shared/build/crypto/Ecdh1PU.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ecdh1PU = void 0;
const ariesAskar_1 = require("../ariesAskar");
const Key_1 = require("./Key");
// Tests
class Ecdh1PU {
    constructor({ apv, apu, algId }) {
        this.algId = algId;
        this.apu = apu;
        this.apv = apv;
    }
    deriveKey({ encAlg, ephemeralKey, recipientKey, senderKey, receive, ccTag, }) {
        return new Key_1.Key(ariesAskar_1.ariesAskar.keyDeriveEcdh1pu({
            algId: this.algId,
            receive,
            apv: this.apv,
            apu: this.apu,
            algorithm: encAlg,
            ephemeralKey: ephemeralKey,
            recipientKey: recipientKey,
            senderKey: senderKey,
            ccTag,
        }));
    }
    encryptDirect({ encAlg, recipientKey, ephemeralKey, senderKey, message, aad, nonce, }) {
        const derived = this.deriveKey({ encAlg, ephemeralKey, recipientKey, senderKey, receive: false });
        const encryptedBuffer = derived.aeadEncrypt({ message, aad, nonce });
        derived.handle.free();
        return encryptedBuffer;
    }
    decryptDirect({ nonce, encAlg, recipientKey, ephemeralKey, senderKey, ciphertext, tag, aad, }) {
        const derived = this.deriveKey({ encAlg, ephemeralKey, recipientKey, senderKey, receive: true });
        const encryptedBuffer = derived.aeadDecrypt({ tag, nonce, ciphertext, aad });
        derived.handle.free();
        return encryptedBuffer;
    }
    senderWrapKey({ wrapAlg, ephemeralKey, recipientKey, senderKey, cek, ccTag, }) {
        const derived = this.deriveKey({
            encAlg: wrapAlg,
            ephemeralKey,
            recipientKey,
            senderKey,
            receive: false,
            ccTag,
        });
        const encryptedBuffer = derived.wrapKey({ other: cek });
        derived.handle.free();
        return encryptedBuffer;
    }
    receiverUnwrapKey({ wrapAlg, encAlg, recipientKey, ephemeralKey, senderKey, ciphertext, nonce, tag, ccTag, }) {
        const derived = this.deriveKey({
            encAlg: wrapAlg,
            ephemeralKey,
            recipientKey,
            receive: true,
            senderKey,
            ccTag,
        });
        const encryptedBuffer = derived.unwrapKey({ tag, nonce, ciphertext, algorithm: encAlg });
        derived.handle.free();
        return encryptedBuffer;
    }
}
exports.Ecdh1PU = Ecdh1PU;
