// @credo-ts/core/build/modules/x509/X509ModuleConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.X509ModuleConfig = void 0;
class X509ModuleConfig {
    constructor(options) {
        this.options = options?.trustedCertificates ? { trustedCertificates: [...options.trustedCertificates] } : {};
        this.options.getTrustedCertificatesForVerification = options?.getTrustedCertificatesForVerification;
    }
    get trustedCertificates() {
        return this.options.trustedCertificates;
    }
    get getTrustedCertificatesForVerification() {
        return this.options.getTrustedCertificatesForVerification;
    }
    setTrustedCertificatesForVerification(fn) {
        this.options.getTrustedCertificatesForVerification = fn;
    }
    setTrustedCertificates(trustedCertificates) {
        this.options.trustedCertificates = trustedCertificates ? [...trustedCertificates] : undefined;
    }
    addTrustedCertificate(trustedCertificate) {
        if (!this.options.trustedCertificates) {
            this.options.trustedCertificates = [trustedCertificate];
            return;
        }
        this.options.trustedCertificates.push(trustedCertificate);
    }
}
exports.X509ModuleConfig = X509ModuleConfig;
