// @openwallet-foundation/askar-shared/build/store/Scan.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Scan = void 0;
const askar_1 = require("../askar");
const error_1 = require("../error");
const EntryList_1 = require("./EntryList");
class Scan {
    constructor({ store, category, limit, offset, profile, tagFilter, orderBy, descending, }) {
        this.category = category;
        this.profile = profile;
        this.tagFilter = tagFilter;
        this.offset = offset;
        this.limit = limit;
        this.orderBy = orderBy;
        this.descending = descending;
        this.store = store;
    }
    get handle() {
        return this._handle;
    }
    async forEachRow(cb) {
        var _a;
        if (!this.handle) {
            if (!((_a = this.store) === null || _a === void 0 ? void 0 : _a.handle))
                throw error_1.AskarError.customError({ message: 'Cannot scan from closed store' });
            this._handle = await askar_1.askar.scanStart({
                storeHandle: this.store.handle,
                profile: this.profile,
                category: this.category,
                tagFilter: this.tagFilter,
                limit: this.limit,
                offset: this.offset,
                orderBy: this.orderBy,
                descending: this.descending,
            });
        }
        if (!this._handle) {
            throw error_1.AskarError.customError({ message: 'Handle is not set. Invalid state reached.' });
        }
        try {
            let recordCount = 0;
            // Loop while limit not reached (or no limit specified)
            while (!this.limit || recordCount < this.limit) {
                const listHandle = await askar_1.askar.scanNext({ scanHandle: this._handle });
                if (!listHandle)
                    break;
                this._listHandle = listHandle;
                const list = new EntryList_1.EntryList({ handle: this._listHandle });
                recordCount = recordCount + list.length;
                for (let index = 0; index < list.length; index++) {
                    const entry = list.getEntryByIndex(index);
                    cb(entry);
                }
            }
        }
        finally {
            askar_1.askar.scanFree({ scanHandle: this._handle });
        }
    }
    async fetchAll() {
        const rows = [];
        await this.forEachRow((row) => rows.push(row.toJson()));
        return rows;
    }
}
exports.Scan = Scan;
