// @hyperledger/indy-vdr-shared/build/indyVdr/IndyVdrRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndyVdrRequest = void 0;
const indyVdr_1 = require("../indyVdr/indyVdr");
class IndyVdrRequest {
    constructor(options) {
        const { handle } = options;
        this._handle = handle;
    }
    get handle() {
        return this._handle;
    }
    get body() {
        return indyVdr_1.indyVdr.requestGetBody({ requestHandle: this.handle });
    }
    get signatureInput() {
        return indyVdr_1.indyVdr.requestGetSignatureInput({ requestHandle: this.handle });
    }
    setEndorser(options) {
        indyVdr_1.indyVdr.requestSetEndorser(Object.assign({ requestHandle: this.handle }, options));
    }
    setMultiSignature(options) {
        indyVdr_1.indyVdr.requestSetMultiSignature(Object.assign({ requestHandle: this.handle }, options));
    }
    setSignature(options) {
        indyVdr_1.indyVdr.requestSetSignature(Object.assign({ requestHandle: this.handle }, options));
    }
    setTransactionAuthorAgreementAcceptance(options) {
        indyVdr_1.indyVdr.requestSetTxnAuthorAgreementAcceptance(Object.assign({ requestHandle: this.handle }, options));
    }
    free() {
        indyVdr_1.indyVdr.requestFree({ requestHandle: this.handle });
    }
}
exports.IndyVdrRequest = IndyVdrRequest;
