// @hyperledger/indy-vdr-shared/build/indyVdr/IndyVdrPool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndyVdrPool = void 0;
const indyVdr_1 = require("./indyVdr");
class IndyVdrPool {
    constructor(options) {
        const { handle } = options;
        this._handle = handle;
    }
    get handle() {
        return this._handle;
    }
    get status() {
        return indyVdr_1.indyVdr.poolGetStatus({ poolHandle: this.handle });
    }
    get verifiers() {
        return indyVdr_1.indyVdr.poolGetVerifiers({ poolHandle: this.handle });
    }
    get transactions() {
        return indyVdr_1.indyVdr.poolGetTransactions({ poolHandle: this.handle });
    }
    async refresh() {
        await indyVdr_1.indyVdr.poolRefresh({ poolHandle: this.handle });
    }
    close() {
        indyVdr_1.indyVdr.poolClose({ poolHandle: this.handle });
    }
    async submitAction(request, options = {}) {
        return indyVdr_1.indyVdr.poolSubmitAction(Object.assign({ poolHandle: this.handle, requestHandle: request.handle }, options));
    }
    async submitRequest(request) {
        return indyVdr_1.indyVdr.poolSubmitRequest({
            poolHandle: this.handle,
            requestHandle: request.handle,
        });
    }
}
exports.IndyVdrPool = IndyVdrPool;
