// @credo-ts/core/build/modules/sd-jwt-vc/repository/SdJwtVcRecord.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SdJwtVcRecord = void 0;
const decode_1 = require("@sd-jwt/decode");
const crypto_1 = require("../../../crypto");
const BaseRecord_1 = require("../../../storage/BaseRecord");
const utils_1 = require("../../../utils");
const uuid_1 = require("../../../utils/uuid");
class SdJwtVcRecord extends BaseRecord_1.BaseRecord {
    // TODO: should we also store the pretty claims so it's not needed to
    // re-calculate the hashes each time? I think for now it's fine to re-calculate
    constructor(props) {
        super();
        this.type = SdJwtVcRecord.type;
        if (props) {
            this.id = props.id ?? (0, uuid_1.uuid)();
            this.createdAt = props.createdAt ?? new Date();
            this.compactSdJwtVc = props.compactSdJwtVc;
            this._tags = props.tags ?? {};
        }
    }
    getTags() {
        const sdjwt = (0, decode_1.decodeSdJwtSync)(this.compactSdJwtVc, crypto_1.Hasher.hash);
        const vct = sdjwt.jwt.payload['vct'];
        const sdAlg = sdjwt.jwt.payload['_sd_alg'];
        const alg = sdjwt.jwt.header['alg'];
        return {
            ...this._tags,
            vct,
            sdAlg: sdAlg ?? 'sha-256',
            alg,
        };
    }
    clone() {
        return utils_1.JsonTransformer.fromJSON(utils_1.JsonTransformer.toJSON(this), this.constructor);
    }
}
exports.SdJwtVcRecord = SdJwtVcRecord;
SdJwtVcRecord.type = 'SdJwtVcRecord';
