// @openwallet-foundation/askar-shared/build/crypto/EcdhEs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EcdhEs = void 0;
const askar_1 = require("../askar");
const Jwk_1 = require("./Jwk");
const Key_1 = require("./Key");
class EcdhEs {
    constructor({ apv, apu, algId }) {
        this.algId = algId;
        this.apu = apu;
        this.apv = apv;
    }
    deriveKey({ encryptionAlgorithm, ephemeralKey, recipientKey, receive, }) {
        return new Key_1.Key(askar_1.askar.keyDeriveEcdhEs({
            algId: this.algId,
            receive,
            apv: this.apv,
            apu: this.apu,
            algorithm: encryptionAlgorithm,
            ephemeralKey,
            recipientKey,
        }));
    }
    encryptDirect({ encryptionAlgorithm, recipientKey, ephemeralKey, message, aad, nonce, }) {
        const eKey = ephemeralKey instanceof Jwk_1.Jwk ? Key_1.Key.fromJwk({ jwk: ephemeralKey }) : ephemeralKey;
        const rKey = recipientKey instanceof Jwk_1.Jwk ? Key_1.Key.fromJwk({ jwk: recipientKey }) : recipientKey;
        const derived = this.deriveKey({ encryptionAlgorithm, ephemeralKey: eKey, recipientKey: rKey, receive: false });
        const encryptedBuffer = derived.aeadEncrypt({ message, aad, nonce });
        derived.handle.free();
        return encryptedBuffer;
    }
    decryptDirect({ nonce, encryptionAlgorithm, recipientKey, ciphertext, ephemeralKey, tag, aad, }) {
        const eKey = ephemeralKey instanceof Jwk_1.Jwk ? Key_1.Key.fromJwk({ jwk: ephemeralKey }) : ephemeralKey;
        const rKey = recipientKey instanceof Jwk_1.Jwk ? Key_1.Key.fromJwk({ jwk: recipientKey }) : recipientKey;
        const derived = this.deriveKey({ encryptionAlgorithm, ephemeralKey: eKey, recipientKey: rKey, receive: true });
        const encryptedBuffer = derived.aeadDecrypt({ tag, nonce, ciphertext, aad });
        derived.handle.free();
        return encryptedBuffer;
    }
    senderWrapKey({ keyWrappingAlgorithm, ephemeralKey, recipientKey, cek, }) {
        const derived = this.deriveKey({
            encryptionAlgorithm: keyWrappingAlgorithm,
            ephemeralKey,
            recipientKey,
            receive: false,
        });
        const encryptedBuffer = derived.wrapKey({ other: cek });
        derived.handle.free();
        return encryptedBuffer;
    }
    receiverUnwrapKey({ recipientKey, keyWrappingAlgorithm, ephemeralKey, encryptionAlgorithm, ciphertext, nonce, tag, }) {
        const derived = this.deriveKey({
            encryptionAlgorithm: keyWrappingAlgorithm,
            ephemeralKey,
            recipientKey,
            receive: true,
        });
        const encryptedBuffer = derived.unwrapKey({ tag, nonce, ciphertext, algorithm: encryptionAlgorithm });
        derived.handle.free();
        return encryptedBuffer;
    }
}
exports.EcdhEs = EcdhEs;
