// @openwallet-foundation/askar-shared/build/crypto/Ecdh1PU.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ecdh1PU = void 0;
const askar_1 = require("../askar");
const Key_1 = require("./Key");
class Ecdh1PU {
    constructor({ apv, apu, algId }) {
        this.algId = algId;
        this.apu = apu;
        this.apv = apv;
    }
    deriveKey({ encryptionAlgorithm, ephemeralKey, recipientKey, senderKey, receive, ccTag, }) {
        return new Key_1.Key(askar_1.askar.keyDeriveEcdh1pu({
            algId: this.algId,
            receive,
            apv: this.apv,
            apu: this.apu,
            algorithm: encryptionAlgorithm,
            ephemeralKey: ephemeralKey,
            recipientKey: recipientKey,
            senderKey: senderKey,
            ccTag,
        }));
    }
    encryptDirect({ encryptionAlgorithm, recipientKey, ephemeralKey, senderKey, message, aad, nonce, }) {
        const derived = this.deriveKey({
            encryptionAlgorithm: encryptionAlgorithm,
            ephemeralKey,
            recipientKey,
            senderKey,
            receive: false,
        });
        const encryptedBuffer = derived.aeadEncrypt({ message, aad, nonce });
        derived.handle.free();
        return encryptedBuffer;
    }
    decryptDirect({ nonce, encryptionAlgorithm, recipientKey, ephemeralKey, senderKey, ciphertext, tag, aad, }) {
        const derived = this.deriveKey({
            encryptionAlgorithm: encryptionAlgorithm,
            ephemeralKey,
            recipientKey,
            senderKey,
            receive: true,
        });
        const encryptedBuffer = derived.aeadDecrypt({ tag, nonce, ciphertext, aad });
        derived.handle.free();
        return encryptedBuffer;
    }
    senderWrapKey({ keyWrappingAlgorithm, ephemeralKey, recipientKey, senderKey, cek, ccTag, }) {
        const derived = this.deriveKey({
            encryptionAlgorithm: keyWrappingAlgorithm,
            ephemeralKey,
            recipientKey,
            senderKey,
            receive: false,
            ccTag,
        });
        const encryptedBuffer = derived.wrapKey({ other: cek });
        derived.handle.free();
        return encryptedBuffer;
    }
    receiverUnwrapKey({ keyWrappingAlgorithm, encryptionAlgorithm, recipientKey, ephemeralKey, senderKey, ciphertext, nonce, tag, ccTag, }) {
        const derived = this.deriveKey({
            encryptionAlgorithm: keyWrappingAlgorithm,
            ephemeralKey,
            recipientKey,
            receive: true,
            senderKey,
            ccTag,
        });
        const encryptedBuffer = derived.unwrapKey({ tag, nonce, ciphertext, algorithm: encryptionAlgorithm });
        derived.handle.free();
        return encryptedBuffer;
    }
}
exports.Ecdh1PU = Ecdh1PU;
