// @openwallet-foundation/askar-shared/build/crypto/handles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalKeyHandle = exports.KeyEntryListHandle = exports.EntryListHandle = exports.SessionHandle = exports.ScanHandle = exports.StoreHandle = exports.ArcHandle = void 0;
const askar_1 = require("../askar");
const error_1 = require("../error");
class ArcHandle {
    constructor(handle) {
        if (handle === '0') {
            throw error_1.AskarError.customError({
                message: 'Invalid handle. This means that the function call succeeded but none was found.',
            });
        }
        this.handle = handle;
    }
    static fromHandle(handle) {
        return fromPointerHandle(ArcHandle, handle);
    }
}
exports.ArcHandle = ArcHandle;
class StoreHandle {
    constructor(handle) {
        this.handle = handle;
    }
    async close() {
        await askar_1.askar.storeClose({ storeHandle: this });
    }
    static fromHandle(handle) {
        return fromSequenceHandle(StoreHandle, handle);
    }
}
exports.StoreHandle = StoreHandle;
class ScanHandle {
    constructor(handle) {
        this.handle = handle;
    }
    free() {
        askar_1.askar.scanFree({ scanHandle: this });
    }
    static fromHandle(handle) {
        return fromSequenceHandle(ScanHandle, handle);
    }
}
exports.ScanHandle = ScanHandle;
class SessionHandle {
    constructor(handle) {
        this.handle = handle;
    }
    async close(commit) {
        await askar_1.askar.sessionClose({ commit, sessionHandle: this });
    }
    static fromHandle(handle) {
        return fromSequenceHandle(SessionHandle, handle);
    }
}
exports.SessionHandle = SessionHandle;
class EntryListHandle extends ArcHandle {
    getCategory(index) {
        return askar_1.askar.entryListGetCategory({ index, entryListHandle: this });
    }
    getName(index) {
        return askar_1.askar.entryListGetName({ index, entryListHandle: this });
    }
    getValue(index) {
        return askar_1.askar.entryListGetValue({ index, entryListHandle: this });
    }
    getTags(index) {
        return askar_1.askar.entryListGetTags({ index, entryListHandle: this });
    }
    free() {
        askar_1.askar.entryListFree({ entryListHandle: this });
    }
    static fromHandle(handle) {
        return fromPointerHandle(EntryListHandle, handle);
    }
}
exports.EntryListHandle = EntryListHandle;
class KeyEntryListHandle extends ArcHandle {
    getAlgorithm(index) {
        return askar_1.askar.keyEntryListGetAlgorithm({ index, keyEntryListHandle: this });
    }
    getName(index) {
        return askar_1.askar.keyEntryListGetName({ index, keyEntryListHandle: this });
    }
    getTags(index) {
        return askar_1.askar.keyEntryListGetTags({ index, keyEntryListHandle: this });
    }
    getMetadata(index) {
        return askar_1.askar.keyEntryListGetMetadata({ index, keyEntryListHandle: this });
    }
    loadKey(index) {
        return askar_1.askar.keyEntryListLoadLocal({ index, keyEntryListHandle: this });
    }
    free() {
        askar_1.askar.keyEntryListFree({ keyEntryListHandle: this });
    }
    static fromHandle(handle) {
        return fromPointerHandle(KeyEntryListHandle, handle);
    }
}
exports.KeyEntryListHandle = KeyEntryListHandle;
class LocalKeyHandle extends ArcHandle {
    free() {
        askar_1.askar.keyFree({ localKeyHandle: this });
    }
    static fromHandle(handle) {
        return fromPointerHandle(LocalKeyHandle, handle);
    }
}
exports.LocalKeyHandle = LocalKeyHandle;
/**
 * Instantiate an handle class based on a received handle. If the handle has a value
 * of null, the handle class won't be instantiated but rather null will be returned.
 */
function fromPointerHandle(HandleClass, handle) {
    return (handle ? new HandleClass(handle) : null);
}
function fromSequenceHandle(HandleClass, handle) {
    if (handle === null) {
        throw error_1.AskarError.customError({ message: 'Invalid handle' });
    }
    return new HandleClass(handle);
}
