// @openwallet-foundation/askar-shared/build/crypto/Key.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Key = void 0;
const buffer_1 = require("buffer");
const askar_1 = require("../askar");
const enums_1 = require("../enums");
const Jwk_1 = require("./Jwk");
class Key {
    constructor(handle) {
        this.localKeyHandle = handle;
    }
    static generate(algorithm, keyBackend, ephemeral = false) {
        return new Key(askar_1.askar.keyGenerate({ algorithm, keyBackend, ephemeral }));
    }
    static fromSeed({ method = enums_1.KeyMethod.None, algorithm, seed, }) {
        return new Key(askar_1.askar.keyFromSeed({ algorithm, method, seed }));
    }
    static fromSecretBytes(options) {
        return new Key(askar_1.askar.keyFromSecretBytes(options));
    }
    static fromPublicBytes(options) {
        return new Key(askar_1.askar.keyFromPublicBytes(options));
    }
    static fromJwk(options) {
        return new Key(askar_1.askar.keyFromJwk(options));
    }
    convertkey(options) {
        return new Key(askar_1.askar.keyConvert(Object.assign({ localKeyHandle: this.handle }, options)));
    }
    keyFromKeyExchange({ algorithm, publicKey }) {
        return new Key(askar_1.askar.keyFromKeyExchange({
            skHandle: this.handle,
            pkHandle: publicKey.handle,
            algorithm,
        }));
    }
    get handle() {
        return this.localKeyHandle;
    }
    get algorithm() {
        const alg = askar_1.askar.keyGetAlgorithm({ localKeyHandle: this.handle });
        return (0, enums_1.keyAlgorithmFromString)(alg);
    }
    get ephemeral() {
        return Boolean(askar_1.askar.keyGetEphemeral({ localKeyHandle: this.handle }));
    }
    get publicBytes() {
        return askar_1.askar.keyGetPublicBytes({ localKeyHandle: this.handle });
    }
    get secretBytes() {
        return askar_1.askar.keyGetSecretBytes({ localKeyHandle: this.handle });
    }
    get jwkPublic() {
        return Jwk_1.Jwk.fromString(askar_1.askar.keyGetJwkPublic({
            localKeyHandle: this.handle,
            algorithm: this.algorithm,
        }));
    }
    get jwkSecret() {
        const secretBytes = askar_1.askar.keyGetJwkSecret({
            localKeyHandle: this.handle,
        });
        return Jwk_1.Jwk.fromString(buffer_1.Buffer.from(secretBytes).toString());
    }
    get jwkThumbprint() {
        return askar_1.askar.keyGetJwkThumbprint({
            localKeyHandle: this.handle,
            algorithm: this.algorithm,
        });
    }
    get aeadParams() {
        return askar_1.askar.keyAeadGetParams({ localKeyHandle: this.handle });
    }
    get aeadRandomNonce() {
        return askar_1.askar.keyAeadRandomNonce({ localKeyHandle: this.handle });
    }
    aeadEncrypt(options) {
        return askar_1.askar.keyAeadEncrypt(Object.assign({ localKeyHandle: this.handle }, options));
    }
    aeadDecrypt(options) {
        return askar_1.askar.keyAeadDecrypt(Object.assign({ localKeyHandle: this.handle }, options));
    }
    signMessage(options) {
        return askar_1.askar.keySignMessage(Object.assign({ localKeyHandle: this.handle }, options));
    }
    verifySignature(options) {
        return askar_1.askar.keyVerifySignature(Object.assign({ localKeyHandle: this.handle }, options));
    }
    wrapKey({ other, nonce }) {
        return askar_1.askar.keyWrapKey({
            localKeyHandle: this.handle,
            other: other.handle,
            nonce,
        });
    }
    unwrapKey(options) {
        return new Key(askar_1.askar.keyUnwrapKey(Object.assign({ localKeyHandle: this.handle }, options)));
    }
}
exports.Key = Key;
