// @credo-ts/core/build/modules/mdoc/repository/MdocRecord.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MdocRecord = void 0;
const BaseRecord_1 = require("../../../storage/BaseRecord");
const utils_1 = require("../../../utils");
const uuid_1 = require("../../../utils/uuid");
const Mdoc_1 = require("../Mdoc");
class MdocRecord extends BaseRecord_1.BaseRecord {
    constructor(props) {
        super();
        this.type = MdocRecord.type;
        if (props) {
            this.id = props.id ?? (0, uuid_1.uuid)();
            this.createdAt = props.createdAt ?? new Date();
            this.base64Url = props.mdoc.base64Url;
            this._tags = props.tags ?? {};
        }
    }
    getTags() {
        const mdoc = Mdoc_1.Mdoc.fromBase64Url(this.base64Url);
        const docType = mdoc.docType;
        const alg = mdoc.alg;
        return {
            ...this._tags,
            docType,
            alg,
        };
    }
    clone() {
        return utils_1.JsonTransformer.fromJSON(utils_1.JsonTransformer.toJSON(this), this.constructor);
    }
}
exports.MdocRecord = MdocRecord;
MdocRecord.type = 'MdocRecord';
