// @sphereon/kmp-mdl-mdoc/encoding-library-base64.js
(function (factory) {
  if (typeof define === 'function' && define.amd)
    define(['exports', './kotlin-kotlin-stdlib.js', './encoding-library-core.js'], factory);
  else if (typeof exports === 'object')
    factory(module.exports, require('./kotlin-kotlin-stdlib.js'), require('./encoding-library-core.js'));
  else {
    if (typeof globalThis['kotlin-kotlin-stdlib'] === 'undefined') {
      throw new Error("Error loading module 'encoding-library-base64'. Its dependency 'kotlin-kotlin-stdlib' was not found. Please, check whether 'kotlin-kotlin-stdlib' is loaded prior to 'encoding-library-base64'.");
    }
    if (typeof globalThis['encoding-library-core'] === 'undefined') {
      throw new Error("Error loading module 'encoding-library-base64'. Its dependency 'encoding-library-core' was not found. Please, check whether 'encoding-library-core' is loaded prior to 'encoding-library-base64'.");
    }
    globalThis['encoding-library-base64'] = factory(typeof globalThis['encoding-library-base64'] === 'undefined' ? {} : globalThis['encoding-library-base64'], globalThis['kotlin-kotlin-stdlib'], globalThis['encoding-library-core']);
  }
}(function (_, kotlin_kotlin, kotlin_io_matthewnelson_encoding_core) {
  'use strict';
  //region block: imports
  var protoOf = kotlin_kotlin.$_$.ia;
  var initMetadataForCompanion = kotlin_kotlin.$_$.j9;
  var equals = kotlin_kotlin.$_$.b9;
  var FunctionAdapter = kotlin_kotlin.$_$.p8;
  var isInterface = kotlin_kotlin.$_$.w9;
  var Flush = kotlin_io_matthewnelson_encoding_core.$_$.f;
  var hashCode = kotlin_kotlin.$_$.h9;
  var initMetadataForClass = kotlin_kotlin.$_$.i9;
  var VOID = kotlin_kotlin.$_$.e;
  var Finalize = kotlin_io_matthewnelson_encoding_core.$_$.e;
  var Unit_instance = kotlin_kotlin.$_$.k4;
  var toByte = kotlin_kotlin.$_$.ja;
  var Companion_instance = kotlin_io_matthewnelson_encoding_core.$_$.b;
  var charSequenceGet = kotlin_kotlin.$_$.v8;
  var Char = kotlin_kotlin.$_$.wc;
  var Config = kotlin_io_matthewnelson_encoding_core.$_$.k;
  var _Char___init__impl__6a9atx = kotlin_kotlin.$_$.y1;
  var Long = kotlin_kotlin.$_$.gd;
  var toLong = kotlin_kotlin.$_$.ka;
  var LinkedHashSet_init_$Create$ = kotlin_kotlin.$_$.s;
  var Setting = kotlin_io_matthewnelson_encoding_core.$_$.j;
  var EncoderDecoder = kotlin_io_matthewnelson_encoding_core.$_$.l;
  var initMetadataForObject = kotlin_kotlin.$_$.n9;
  var FeedBuffer = kotlin_io_matthewnelson_encoding_core.$_$.g;
  var Feed = kotlin_io_matthewnelson_encoding_core.$_$.h;
  var toString = kotlin_kotlin.$_$.b2;
  var EncodingException_init_$Create$ = kotlin_io_matthewnelson_encoding_core.$_$.a;
  var Char__toInt_impl_vasixd = kotlin_kotlin.$_$.a2;
  var Feed_0 = kotlin_io_matthewnelson_encoding_core.$_$.i;
  var objectCreate = kotlin_kotlin.$_$.ha;
  //endregion
  //region block: pre-declaration
  initMetadataForCompanion(Companion);
  initMetadataForClass(sam$io_matthewnelson_encoding_core_util_FeedBuffer_Flush$0, 'sam$io_matthewnelson_encoding_core_util_FeedBuffer_Flush$0', VOID, VOID, [Flush, FunctionAdapter]);
  initMetadataForClass(sam$io_matthewnelson_encoding_core_util_FeedBuffer_Finalize$0, 'sam$io_matthewnelson_encoding_core_util_FeedBuffer_Finalize$0', VOID, VOID, [Finalize, FunctionAdapter]);
  initMetadataForClass(sam$io_matthewnelson_encoding_core_util_FeedBuffer_Flush$0_0, 'sam$io_matthewnelson_encoding_core_util_FeedBuffer_Flush$0', VOID, VOID, [Flush, FunctionAdapter]);
  initMetadataForClass(sam$io_matthewnelson_encoding_core_util_FeedBuffer_Finalize$0_0, 'sam$io_matthewnelson_encoding_core_util_FeedBuffer_Finalize$0', VOID, VOID, [Finalize, FunctionAdapter]);
  initMetadataForClass(Config_0, 'Config', VOID, Config);
  initMetadataForObject(Default, 'Default', VOID, EncoderDecoder);
  initMetadataForObject(UrlSafe, 'UrlSafe', VOID, EncoderDecoder);
  initMetadataForClass(DecodingBuffer, 'DecodingBuffer', VOID, FeedBuffer);
  initMetadataForClass(EncodingBuffer, 'EncodingBuffer', VOID, FeedBuffer);
  initMetadataForClass(Base64$newDecoderFeedProtected$1, VOID, VOID, Feed);
  initMetadataForClass(Base64$newEncoderFeedProtected$1, VOID, VOID, Feed_0);
  initMetadataForClass(Base64, 'Base64', VOID, EncoderDecoder);
  initMetadataForClass(Base64ConfigBuilder, 'Base64ConfigBuilder', Base64ConfigBuilder_init_$Create$);
  //endregion
  function Companion() {
  }
  protoOf(Companion).j2f = function (builder) {
    return new Config_0(builder.k2f_1, builder.l2f_1, builder.m2f_1, builder.n2f_1);
  };
  var Companion_instance_0;
  function Companion_getInstance() {
    return Companion_instance_0;
  }
  function sam$io_matthewnelson_encoding_core_util_FeedBuffer_Flush$0(function_0) {
    this.o2f_1 = function_0;
  }
  protoOf(sam$io_matthewnelson_encoding_core_util_FeedBuffer_Flush$0).w2e = function (buffer) {
    return this.o2f_1(buffer);
  };
  protoOf(sam$io_matthewnelson_encoding_core_util_FeedBuffer_Flush$0).v3 = function () {
    return this.o2f_1;
  };
  protoOf(sam$io_matthewnelson_encoding_core_util_FeedBuffer_Flush$0).equals = function (other) {
    var tmp;
    if (!(other == null) ? isInterface(other, Flush) : false) {
      var tmp_0;
      if (!(other == null) ? isInterface(other, FunctionAdapter) : false) {
        tmp_0 = equals(this.v3(), other.v3());
      } else {
        tmp_0 = false;
      }
      tmp = tmp_0;
    } else {
      tmp = false;
    }
    return tmp;
  };
  protoOf(sam$io_matthewnelson_encoding_core_util_FeedBuffer_Flush$0).hashCode = function () {
    return hashCode(this.v3());
  };
  function sam$io_matthewnelson_encoding_core_util_FeedBuffer_Finalize$0(function_0) {
    this.p2f_1 = function_0;
  }
  protoOf(sam$io_matthewnelson_encoding_core_util_FeedBuffer_Finalize$0).x2e = function (modulus, buffer) {
    return this.p2f_1(modulus, buffer);
  };
  protoOf(sam$io_matthewnelson_encoding_core_util_FeedBuffer_Finalize$0).v3 = function () {
    return this.p2f_1;
  };
  protoOf(sam$io_matthewnelson_encoding_core_util_FeedBuffer_Finalize$0).equals = function (other) {
    var tmp;
    if (!(other == null) ? isInterface(other, Finalize) : false) {
      var tmp_0;
      if (!(other == null) ? isInterface(other, FunctionAdapter) : false) {
        tmp_0 = equals(this.v3(), other.v3());
      } else {
        tmp_0 = false;
      }
      tmp = tmp_0;
    } else {
      tmp = false;
    }
    return tmp;
  };
  protoOf(sam$io_matthewnelson_encoding_core_util_FeedBuffer_Finalize$0).hashCode = function () {
    return hashCode(this.v3());
  };
  function Base64$DecodingBuffer$_init_$lambda_ohvg0l($out) {
    return function (buffer) {
      var bitBuffer = 0;
      var inductionVariable = 0;
      var last = buffer.length;
      while (inductionVariable < last) {
        var bits = buffer[inductionVariable];
        inductionVariable = inductionVariable + 1 | 0;
        bitBuffer = bitBuffer << 6 | bits;
      }
      $out.d2d(toByte(bitBuffer >> 16));
      $out.d2d(toByte(bitBuffer >> 8));
      $out.d2d(toByte(bitBuffer));
      return Unit_instance;
    };
  }
  function Base64$DecodingBuffer$_init_$lambda_ohvg0l_0($out) {
    return function (modulus, buffer) {
      var tmp;
      if (modulus === 1) {
        throw Companion_instance.y2e(modulus);
      }
      var bitBuffer = 0;
      var inductionVariable = 0;
      var tmp_0;
      if (inductionVariable < modulus) {
        do {
          var i = inductionVariable;
          inductionVariable = inductionVariable + 1 | 0;
          bitBuffer = bitBuffer << 6 | buffer[i];
        }
         while (inductionVariable < modulus);
        tmp_0 = Unit_instance;
      }
      var tmp_1;
      switch (modulus) {
        case 0:
          tmp_1 = Unit_instance;
          break;
        case 2:
          bitBuffer = bitBuffer << 12;
          $out.d2d(toByte(bitBuffer >> 16));
          tmp_1 = Unit_instance;
          break;
        case 3:
          bitBuffer = bitBuffer << 6;
          $out.d2d(toByte(bitBuffer >> 16));
          $out.d2d(toByte(bitBuffer >> 8));
          tmp_1 = Unit_instance;
          break;
      }
      return Unit_instance;
    };
  }
  function sam$io_matthewnelson_encoding_core_util_FeedBuffer_Flush$0_0(function_0) {
    this.q2f_1 = function_0;
  }
  protoOf(sam$io_matthewnelson_encoding_core_util_FeedBuffer_Flush$0_0).w2e = function (buffer) {
    return this.q2f_1(buffer);
  };
  protoOf(sam$io_matthewnelson_encoding_core_util_FeedBuffer_Flush$0_0).v3 = function () {
    return this.q2f_1;
  };
  protoOf(sam$io_matthewnelson_encoding_core_util_FeedBuffer_Flush$0_0).equals = function (other) {
    var tmp;
    if (!(other == null) ? isInterface(other, Flush) : false) {
      var tmp_0;
      if (!(other == null) ? isInterface(other, FunctionAdapter) : false) {
        tmp_0 = equals(this.v3(), other.v3());
      } else {
        tmp_0 = false;
      }
      tmp = tmp_0;
    } else {
      tmp = false;
    }
    return tmp;
  };
  protoOf(sam$io_matthewnelson_encoding_core_util_FeedBuffer_Flush$0_0).hashCode = function () {
    return hashCode(this.v3());
  };
  function sam$io_matthewnelson_encoding_core_util_FeedBuffer_Finalize$0_0(function_0) {
    this.r2f_1 = function_0;
  }
  protoOf(sam$io_matthewnelson_encoding_core_util_FeedBuffer_Finalize$0_0).x2e = function (modulus, buffer) {
    return this.r2f_1(modulus, buffer);
  };
  protoOf(sam$io_matthewnelson_encoding_core_util_FeedBuffer_Finalize$0_0).v3 = function () {
    return this.r2f_1;
  };
  protoOf(sam$io_matthewnelson_encoding_core_util_FeedBuffer_Finalize$0_0).equals = function (other) {
    var tmp;
    if (!(other == null) ? isInterface(other, Finalize) : false) {
      var tmp_0;
      if (!(other == null) ? isInterface(other, FunctionAdapter) : false) {
        tmp_0 = equals(this.v3(), other.v3());
      } else {
        tmp_0 = false;
      }
      tmp = tmp_0;
    } else {
      tmp = false;
    }
    return tmp;
  };
  protoOf(sam$io_matthewnelson_encoding_core_util_FeedBuffer_Finalize$0_0).hashCode = function () {
    return hashCode(this.v3());
  };
  function Base64$EncodingBuffer$_init_$lambda_lcnz7h($out, $table) {
    return function (buffer) {
      var b0 = buffer[0];
      var b1 = buffer[1];
      var b2 = buffer[2];
      $out.y2d(charSequenceGet($table, (b0 & 255) >> 2));
      $out.y2d(charSequenceGet($table, (b0 & 3) << 4 | (b1 & 255) >> 4));
      $out.y2d(charSequenceGet($table, (b1 & 15) << 2 | (b2 & 255) >> 6));
      $out.y2d(charSequenceGet($table, b2 & 63));
      return Unit_instance;
    };
  }
  function Base64$EncodingBuffer$_init_$lambda_lcnz7h_0($out, $table, $paddingChar) {
    return function (modulus, buffer) {
      var tmp;
      switch (modulus) {
        case 0:
          tmp = 0;
          break;
        case 1:
          var b0 = buffer[0];
          $out.y2d(charSequenceGet($table, (b0 & 255) >> 2));
          $out.y2d(charSequenceGet($table, (b0 & 3) << 4));
          tmp = 2;
          break;
        default:
          var b0_0 = buffer[0];
          var b1 = buffer[1];
          $out.y2d(charSequenceGet($table, (b0_0 & 255) >> 2));
          $out.y2d(charSequenceGet($table, (b0_0 & 3) << 4 | (b1 & 255) >> 4));
          $out.y2d(charSequenceGet($table, (b1 & 15) << 2));
          tmp = 1;
          break;
      }
      var padCount = tmp;
      var tmp_0;
      var tmp_1 = $paddingChar;
      if (!((tmp_1 == null ? null : new Char(tmp_1)) == null)) {
        // Inline function 'kotlin.contracts.contract' call
        var inductionVariable = 0;
        var tmp_2;
        if (inductionVariable < padCount) {
          do {
            var index = inductionVariable;
            inductionVariable = inductionVariable + 1 | 0;
            // Inline function 'io.matthewnelson.encoding.base64.EncodingBuffer.<init>.<anonymous>.<anonymous>' call
            $out.y2d($paddingChar);
          }
           while (inductionVariable < padCount);
          tmp_2 = Unit_instance;
        }
        tmp_0 = tmp_2;
      }
      return Unit_instance;
    };
  }
  function Config_0(isLenient, lineBreakInterval, encodeToUrlSafe, padEncoded) {
    Config.call(this, isLenient, lineBreakInterval, _Char___init__impl__6a9atx(61));
    this.v2f_1 = encodeToUrlSafe;
    this.w2f_1 = padEncoded;
  }
  protoOf(Config_0).x2f = function (encodedSize) {
    return encodedSize.f3(new Long(6, 0)).g3(new Long(8, 0));
  };
  protoOf(Config_0).r2e = function (encodedSize, input) {
    return this.x2f(toLong(encodedSize)).i1();
  };
  protoOf(Config_0).o2e = function (unEncodedSize) {
    var outSize = unEncodedSize.d3(new Long(2, 0)).g3(new Long(3, 0)).f3(new Long(4, 0));
    if (this.w2f_1)
      return outSize;
    // Inline function 'kotlin.Long.rem' call
    var tmp$ret$0 = unEncodedSize.h3(toLong(3));
    var tmp0_subject = unEncodedSize.e3(unEncodedSize.e3(tmp$ret$0));
    if (!tmp0_subject.equals(new Long(0, 0)))
      if (tmp0_subject.equals(new Long(1, 0)))
        outSize = outSize.e3(new Long(2, 0));
      else if (tmp0_subject.equals(new Long(2, 0)))
        outSize = outSize.e3(new Long(1, 0));
    return outSize;
  };
  protoOf(Config_0).s2e = function () {
    // Inline function 'kotlin.collections.buildSet' call
    // Inline function 'kotlin.contracts.contract' call
    // Inline function 'kotlin.collections.buildSetInternal' call
    // Inline function 'kotlin.apply' call
    var this_0 = LinkedHashSet_init_$Create$();
    // Inline function 'kotlin.contracts.contract' call
    // Inline function 'io.matthewnelson.encoding.base64.Config.toStringAddSettings.<anonymous>' call
    this_0.l(new Setting(this, 'encodeToUrlSafe', this.v2f_1));
    this_0.l(new Setting(this, 'padEncoded', this.w2f_1));
    return this_0.y3();
  };
  function Default() {
    Default_instance = this;
    // Inline function 'kotlin.apply' call
    var this_0 = Base64ConfigBuilder_init_$Create$();
    // Inline function 'kotlin.contracts.contract' call
    // Inline function 'io.matthewnelson.encoding.base64.Default.<init>.<anonymous>' call
    this_0.l2f_1 = 64;
    EncoderDecoder.call(this, this_0.q19());
    this.z2f_1 = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
    this.a2g_1 = new Base64(this.i2d_1);
  }
  protoOf(Default).u2e = function () {
    return this.a2g_1.u2e();
  };
  protoOf(Default).w2d = function (out) {
    return this.a2g_1.w2d(out);
  };
  protoOf(Default).j2e = function (out) {
    return this.a2g_1.j2e(out);
  };
  var Default_instance;
  function Default_getInstance() {
    if (Default_instance == null)
      new Default();
    return Default_instance;
  }
  function UrlSafe() {
    UrlSafe_instance = this;
    // Inline function 'kotlin.apply' call
    var this_0 = Base64ConfigBuilder_init_$Create$_0(Default_getInstance().i2d_1);
    // Inline function 'kotlin.contracts.contract' call
    // Inline function 'io.matthewnelson.encoding.base64.UrlSafe.<init>.<anonymous>' call
    this_0.m2f_1 = true;
    EncoderDecoder.call(this, this_0.q19());
    this.d2g_1 = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_';
    this.e2g_1 = new Base64(this.i2d_1);
  }
  protoOf(UrlSafe).u2e = function () {
    return this.e2g_1.u2e();
  };
  protoOf(UrlSafe).w2d = function (out) {
    return this.e2g_1.w2d(out);
  };
  protoOf(UrlSafe).j2e = function (out) {
    return this.e2g_1.j2e(out);
  };
  var UrlSafe_instance;
  function UrlSafe_getInstance() {
    if (UrlSafe_instance == null)
      new UrlSafe();
    return UrlSafe_instance;
  }
  function DecodingBuffer($outer, out) {
    this.k2g_1 = $outer;
    var tmp = Base64$DecodingBuffer$_init_$lambda_ohvg0l(out);
    var tmp_0 = new sam$io_matthewnelson_encoding_core_util_FeedBuffer_Flush$0(tmp);
    var tmp_1 = Base64$DecodingBuffer$_init_$lambda_ohvg0l_0(out);
    FeedBuffer.call(this, 4, tmp_0, new sam$io_matthewnelson_encoding_core_util_FeedBuffer_Finalize$0(tmp_1));
  }
  function EncodingBuffer($outer, out, table, paddingChar) {
    this.q2g_1 = $outer;
    var tmp = Base64$EncodingBuffer$_init_$lambda_lcnz7h(out, table);
    var tmp_0 = new sam$io_matthewnelson_encoding_core_util_FeedBuffer_Flush$0_0(tmp);
    var tmp_1 = Base64$EncodingBuffer$_init_$lambda_lcnz7h_0(out, table, paddingChar);
    FeedBuffer.call(this, 3, tmp_0, new sam$io_matthewnelson_encoding_core_util_FeedBuffer_Finalize$0_0(tmp_1));
  }
  function Base64$newDecoderFeedProtected$1(this$0, $out) {
    Feed.call(this, this$0);
    this.v2g_1 = new DecodingBuffer(this$0, $out);
  }
  protoOf(Base64$newDecoderFeedProtected$1).o2d = function (input) {
    var tmp;
    if (_Char___init__impl__6a9atx(48) <= input ? input <= _Char___init__impl__6a9atx(57) : false) {
      // Inline function 'kotlin.code' call
      tmp = Char__toInt_impl_vasixd(input) + 4 | 0;
    } else if (_Char___init__impl__6a9atx(65) <= input ? input <= _Char___init__impl__6a9atx(90) : false) {
      // Inline function 'kotlin.code' call
      tmp = Char__toInt_impl_vasixd(input) - 65 | 0;
    } else if (_Char___init__impl__6a9atx(97) <= input ? input <= _Char___init__impl__6a9atx(122) : false) {
      // Inline function 'kotlin.code' call
      tmp = Char__toInt_impl_vasixd(input) - 71 | 0;
    } else if (input === _Char___init__impl__6a9atx(43) ? true : input === _Char___init__impl__6a9atx(45)) {
      tmp = 62;
    } else if (input === _Char___init__impl__6a9atx(47) ? true : input === _Char___init__impl__6a9atx(95)) {
      tmp = 63;
    } else {
      throw EncodingException_init_$Create$('Char[' + toString(input) + '] is not a valid Base64 character');
    }
    var bits = tmp;
    this.v2g_1.e2f(bits);
  };
  protoOf(Base64$newDecoderFeedProtected$1).s2d = function () {
    this.v2g_1.f2f();
  };
  function Base64$newEncoderFeedProtected$1(this$0, $out) {
    Feed_0.call(this, this$0);
    var tmp = this;
    var tmp_0;
    if (this.n2d_1.v2f_1) {
      tmp_0 = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_';
    } else {
      tmp_0 = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
    }
    var tmp_1 = tmp_0;
    var tmp_2;
    if (this.n2d_1.w2f_1) {
      tmp_2 = this.n2d_1.l2d_1;
    } else {
      tmp_2 = null;
    }
    tmp.z2g_1 = new EncodingBuffer(this$0, $out, tmp_1, tmp_2);
  }
  protoOf(Base64$newEncoderFeedProtected$1).d2e = function (input) {
    this.z2g_1.e2f(input);
  };
  protoOf(Base64$newEncoderFeedProtected$1).s2d = function () {
    this.z2g_1.f2f();
  };
  function Base64(config) {
    EncoderDecoder.call(this, config);
  }
  protoOf(Base64).w2d = function (out) {
    return new Base64$newDecoderFeedProtected$1(this, out);
  };
  protoOf(Base64).j2e = function (out) {
    return new Base64$newEncoderFeedProtected$1(this, out);
  };
  protoOf(Base64).u2e = function () {
    return 'Base64';
  };
  function Base64_0(block) {
    return Base64_1(null, block);
  }
  function Base64ConfigBuilder_init_$Init$($this) {
    Base64ConfigBuilder.call($this);
    return $this;
  }
  function Base64ConfigBuilder_init_$Create$() {
    return Base64ConfigBuilder_init_$Init$(objectCreate(protoOf(Base64ConfigBuilder)));
  }
  function Base64ConfigBuilder_init_$Init$_0(config, $this) {
    Base64ConfigBuilder_init_$Init$($this);
    if (config == null)
      return $this;
    var tmp = $this;
    var tmp0_elvis_lhs = config.k2d_1;
    tmp.k2f_1 = tmp0_elvis_lhs == null ? true : tmp0_elvis_lhs;
    $this.l2f_1 = config.m2d_1;
    $this.m2f_1 = config.v2f_1;
    $this.n2f_1 = config.w2f_1;
    return $this;
  }
  function Base64ConfigBuilder_init_$Create$_0(config) {
    return Base64ConfigBuilder_init_$Init$_0(config, objectCreate(protoOf(Base64ConfigBuilder)));
  }
  protoOf(Base64ConfigBuilder).q19 = function () {
    return Companion_instance_0.j2f(this);
  };
  function Base64ConfigBuilder() {
    this.k2f_1 = true;
    this.l2f_1 = 0;
    this.m2f_1 = false;
    this.n2f_1 = true;
  }
  function Base64_1(config, block) {
    var builder = Base64ConfigBuilder_init_$Create$_0(config);
    block(builder);
    return new Base64(builder.q19());
  }
  //region block: init
  Companion_instance_0 = new Companion();
  //endregion
  //region block: exports
  _.$_$ = _.$_$ || {};
  _.$_$.a = Default_getInstance;
  _.$_$.b = UrlSafe_getInstance;
  _.$_$.c = Base64_0;
  //endregion
  return _;
}));
