// @openwallet-foundation/askar-shared/build/enums/KeyAlgorithm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.keyAlgorithmToString = exports.keyAlgorithmFromString = exports.KeyAlgorithm = void 0;
const error_1 = require("../error");
var KeyAlgorithm;
(function (KeyAlgorithm) {
    KeyAlgorithm["AesA128Gcm"] = "a128gcm";
    KeyAlgorithm["AesA256Gcm"] = "a256gcm";
    KeyAlgorithm["AesA128CbcHs256"] = "a128cbchs256";
    KeyAlgorithm["AesA256CbcHs512"] = "a256cbchs512";
    KeyAlgorithm["AesA128Kw"] = "a128kw";
    KeyAlgorithm["AesA256Kw"] = "a256kw";
    KeyAlgorithm["Bls12381G1"] = "bls12381g1";
    KeyAlgorithm["Bls12381G2"] = "bls12381g2";
    KeyAlgorithm["Chacha20C20P"] = "c20p";
    KeyAlgorithm["Chacha20XC20P"] = "xc20p";
    KeyAlgorithm["Ed25519"] = "ed25519";
    KeyAlgorithm["X25519"] = "x25519";
    KeyAlgorithm["EcSecp256k1"] = "k256";
    KeyAlgorithm["EcSecp256r1"] = "p256";
    KeyAlgorithm["EcSecp384r1"] = "p384";
})(KeyAlgorithm || (exports.KeyAlgorithm = KeyAlgorithm = {}));
const keyAlgorithmFromString = (alg) => {
    const keyAlg = Object.entries(KeyAlgorithm).find(([, value]) => value === alg);
    if (keyAlg)
        return keyAlg[1];
    throw error_1.AskarError.customError({ message: `Algorithm: ${alg} is not supported!` });
};
exports.keyAlgorithmFromString = keyAlgorithmFromString;
const keyAlgorithmToString = (alg) => {
    const keyAlg = Object.entries(KeyAlgorithm).find(([key]) => key === alg);
    if (keyAlg)
        return keyAlg[0];
    throw error_1.AskarError.customError({ message: `Algorithm: ${alg} is not supported!` });
};
exports.keyAlgorithmToString = keyAlgorithmToString;
