// @animo-id/pex/dist/browser/lib/utils/VCUtils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.filterToRestrictedDIDs = exports.isRestrictedDID = exports.uniformDIDMethods = exports.formatValidationErrors = exports.formatValidationError = exports.definitionVersionDiscovery = exports.getIssuerString = exports.getSubjectIdsAsString = void 0;
const ssi_types_1 = require("@sphereon/ssi-types");
const types_1 = require("../types");
const validatePDv1_js_1 = __importDefault(require("../validation/validatePDv1.js"));
const validatePDv2_js_1 = __importDefault(require("../validation/validatePDv2.js"));
const ObjectUtils_1 = require("./ObjectUtils");
const jsonPathUtils_1 = require("./jsonPathUtils");
function getSubjectIdsAsString(vc) {
    if (ssi_types_1.CredentialMapper.isSdJwtDecodedCredential(vc)) {
        // TODO: should we also handle `cnf` claim?
        return vc.signedPayload.sub ? [vc.signedPayload.sub] : [];
    }
    const subjects = Array.isArray(vc.credentialSubject) ? vc.credentialSubject : [vc.credentialSubject];
    return subjects.filter((s) => !!s.id).map((value) => value.id);
}
exports.getSubjectIdsAsString = getSubjectIdsAsString;
function getIssuerString(vc) {
    if (ssi_types_1.CredentialMapper.isSdJwtDecodedCredential(vc)) {
        return vc.signedPayload.iss;
    }
    return ObjectUtils_1.ObjectUtils.isString(vc.issuer) ? vc.issuer : vc.issuer.id;
}
exports.getIssuerString = getIssuerString;
function definitionVersionDiscovery(presentationDefinition) {
    var _a, _b;
    const presentationDefinitionCopy = JSON.parse(JSON.stringify(presentationDefinition));
    jsonPathUtils_1.JsonPathUtils.changePropertyNameRecursively(presentationDefinitionCopy, '_const', 'const');
    jsonPathUtils_1.JsonPathUtils.changePropertyNameRecursively(presentationDefinitionCopy, '_enum', 'enum');
    const data = { presentation_definition: presentationDefinitionCopy };
    if ((0, validatePDv2_js_1.default)(data)) {
        return { version: types_1.PEVersion.v2 };
    }
    const v2Errors = (_a = validatePDv2_js_1.default.errors) !== null && _a !== void 0 ? _a : undefined;
    if ((0, validatePDv1_js_1.default)(data)) {
        return { version: types_1.PEVersion.v1 };
    }
    const v1Errors = (_b = validatePDv1_js_1.default.errors) !== null && _b !== void 0 ? _b : undefined;
    return {
        error: 'This is not a valid PresentationDefinition',
        v1Errors,
        v2Errors,
    };
}
exports.definitionVersionDiscovery = definitionVersionDiscovery;
function formatValidationError(error) {
    return `${error.instancePath || '/'}: ${error.message}${error.params.additionalProperty ? ` (${error.params.additionalProperty})` : ''}`;
}
exports.formatValidationError = formatValidationError;
function formatValidationErrors(errors) {
    if (!(errors === null || errors === void 0 ? void 0 : errors.length)) {
        return undefined;
    }
    return errors.map(formatValidationError).join('\n  ');
}
exports.formatValidationErrors = formatValidationErrors;
function uniformDIDMethods(dids, opts) {
    var _a;
    let result = (_a = dids === null || dids === void 0 ? void 0 : dids.map((did) => did.toLowerCase()).map((did) => (did.startsWith('did:') ? did : `did:${did}`))) !== null && _a !== void 0 ? _a : [];
    if (opts === null || opts === void 0 ? void 0 : opts.removePrefix) {
        const length = opts.removePrefix.endsWith(':') ? opts.removePrefix.length : opts.removePrefix.length + 1;
        result = result.map((did) => (did.startsWith(opts.removePrefix) ? did.substring(length) : did));
    }
    if (result.includes('did')) {
        // The string did denotes every DID method, hence we return an empty array, indicating all methods are supported
        return [];
    }
    return result;
}
exports.uniformDIDMethods = uniformDIDMethods;
function isRestrictedDID(DID, restrictToDIDMethods) {
    const methods = uniformDIDMethods(restrictToDIDMethods);
    return methods.length === 0 || methods.some((method) => DID.toLowerCase().startsWith(method));
}
exports.isRestrictedDID = isRestrictedDID;
function filterToRestrictedDIDs(DIDs, restrictToDIDMethods) {
    const methods = uniformDIDMethods(restrictToDIDMethods);
    if (methods.length === 0) {
        return DIDs;
    }
    return methods.flatMap((method) => DIDs.filter((DID) => DID.toLowerCase().startsWith(method)));
}
exports.filterToRestrictedDIDs = filterToRestrictedDIDs;
