// @hyperledger/aries-askar-shared/build/enums/KeyAlgs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.keyAlgToString = exports.keyAlgFromString = exports.KeyAlgs = void 0;
const error_1 = require("../error");
var KeyAlgs;
(function (KeyAlgs) {
    KeyAlgs["AesA128Gcm"] = "a128gcm";
    KeyAlgs["AesA256Gcm"] = "a256gcm";
    KeyAlgs["AesA128CbcHs256"] = "a128cbchs256";
    KeyAlgs["AesA256CbcHs512"] = "a256cbchs512";
    KeyAlgs["AesA128Kw"] = "a128kw";
    KeyAlgs["AesA256Kw"] = "a256kw";
    KeyAlgs["Bls12381G1"] = "bls12381g1";
    KeyAlgs["Bls12381G2"] = "bls12381g2";
    KeyAlgs["Bls12381G1G2"] = "bls12381g1g2";
    KeyAlgs["Chacha20C20P"] = "c20p";
    KeyAlgs["Chacha20XC20P"] = "xc20p";
    KeyAlgs["Ed25519"] = "ed25519";
    KeyAlgs["X25519"] = "x25519";
    KeyAlgs["EcSecp256k1"] = "k256";
    KeyAlgs["EcSecp256r1"] = "p256";
    KeyAlgs["EcSecp384r1"] = "p384";
})(KeyAlgs || (exports.KeyAlgs = KeyAlgs = {}));
const keyAlgFromString = (alg) => {
    const keyAlg = Object.entries(KeyAlgs).find(([, value]) => value === alg);
    if (keyAlg)
        return keyAlg[1];
    throw error_1.AriesAskarError.customError({ message: `Algorithm: ${alg} is not supported!` });
};
exports.keyAlgFromString = keyAlgFromString;
const keyAlgToString = (alg) => {
    const keyAlg = Object.entries(KeyAlgs).find(([key]) => key === alg);
    if (keyAlg)
        return keyAlg[0];
    throw error_1.AriesAskarError.customError({ message: `Algorithm: ${alg} is not supported!` });
};
exports.keyAlgToString = keyAlgToString;
