// @sphereon/kmp-mdl-mdoc/@sphereon/kmp-common.js
(function (factory) {
  if (typeof define === 'function' && define.amd)
    define(['exports', '../kotlin-kotlin-stdlib.js', '../kotlinx-serialization-kotlinx-serialization-json.js', '../Kotlin-DateTime-library-kotlinx-datetime.js', '../kotlinx-serialization-kotlinx-serialization-core.js', '../encoding-library-core.js', '../encoding-library-base64.js', '../Kermit-kermit-core.js', '../Kermit-kermit.js'], factory);
  else if (typeof exports === 'object')
    factory(module.exports, require('../kotlin-kotlin-stdlib.js'), require('../kotlinx-serialization-kotlinx-serialization-json.js'), require('../Kotlin-DateTime-library-kotlinx-datetime.js'), require('../kotlinx-serialization-kotlinx-serialization-core.js'), require('../encoding-library-core.js'), require('../encoding-library-base64.js'), require('../Kermit-kermit-core.js'), require('../Kermit-kermit.js'));
  else {
    if (typeof globalThis['kotlin-kotlin-stdlib'] === 'undefined') {
      throw new Error("Error loading module '@sphereon/kmp-common'. Its dependency 'kotlin-kotlin-stdlib' was not found. Please, check whether 'kotlin-kotlin-stdlib' is loaded prior to '@sphereon/kmp-common'.");
    }
    if (typeof globalThis['kotlinx-serialization-kotlinx-serialization-json'] === 'undefined') {
      throw new Error("Error loading module '@sphereon/kmp-common'. Its dependency 'kotlinx-serialization-kotlinx-serialization-json' was not found. Please, check whether 'kotlinx-serialization-kotlinx-serialization-json' is loaded prior to '@sphereon/kmp-common'.");
    }
    if (typeof globalThis['Kotlin-DateTime-library-kotlinx-datetime'] === 'undefined') {
      throw new Error("Error loading module '@sphereon/kmp-common'. Its dependency 'Kotlin-DateTime-library-kotlinx-datetime' was not found. Please, check whether 'Kotlin-DateTime-library-kotlinx-datetime' is loaded prior to '@sphereon/kmp-common'.");
    }
    if (typeof globalThis['kotlinx-serialization-kotlinx-serialization-core'] === 'undefined') {
      throw new Error("Error loading module '@sphereon/kmp-common'. Its dependency 'kotlinx-serialization-kotlinx-serialization-core' was not found. Please, check whether 'kotlinx-serialization-kotlinx-serialization-core' is loaded prior to '@sphereon/kmp-common'.");
    }
    if (typeof globalThis['encoding-library-core'] === 'undefined') {
      throw new Error("Error loading module '@sphereon/kmp-common'. Its dependency 'encoding-library-core' was not found. Please, check whether 'encoding-library-core' is loaded prior to '@sphereon/kmp-common'.");
    }
    if (typeof globalThis['encoding-library-base64'] === 'undefined') {
      throw new Error("Error loading module '@sphereon/kmp-common'. Its dependency 'encoding-library-base64' was not found. Please, check whether 'encoding-library-base64' is loaded prior to '@sphereon/kmp-common'.");
    }
    if (typeof globalThis['Kermit-kermit-core'] === 'undefined') {
      throw new Error("Error loading module '@sphereon/kmp-common'. Its dependency 'Kermit-kermit-core' was not found. Please, check whether 'Kermit-kermit-core' is loaded prior to '@sphereon/kmp-common'.");
    }
    if (typeof globalThis['Kermit-kermit'] === 'undefined') {
      throw new Error("Error loading module '@sphereon/kmp-common'. Its dependency 'Kermit-kermit' was not found. Please, check whether 'Kermit-kermit' is loaded prior to '@sphereon/kmp-common'.");
    }
    globalThis['@sphereon/kmp-common'] = factory(typeof globalThis['@sphereon/kmp-common'] === 'undefined' ? {} : globalThis['@sphereon/kmp-common'], globalThis['kotlin-kotlin-stdlib'], globalThis['kotlinx-serialization-kotlinx-serialization-json'], globalThis['Kotlin-DateTime-library-kotlinx-datetime'], globalThis['kotlinx-serialization-kotlinx-serialization-core'], globalThis['encoding-library-core'], globalThis['encoding-library-base64'], globalThis['Kermit-kermit-core'], globalThis['Kermit-kermit']);
  }
}(function (_, kotlin_kotlin, kotlin_org_jetbrains_kotlinx_kotlinx_serialization_json, kotlin_org_jetbrains_kotlinx_kotlinx_datetime, kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core, kotlin_io_matthewnelson_encoding_core, kotlin_io_matthewnelson_encoding_base64, kotlin_co_touchlab_kermit_core, kotlin_co_touchlab_kermit) {
  'use strict';
  //region block: imports
  var toList = kotlin_kotlin.$_$.l7;
  var toSet = kotlin_kotlin.$_$.s7;
  var LinkedHashMap_init_$Create$ = kotlin_kotlin.$_$.q;
  var toMutableMap = kotlin_kotlin.$_$.q7;
  var Pair = kotlin_kotlin.$_$.jd;
  var mapOf = kotlin_kotlin.$_$.s6;
  var JsonObject = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_json.$_$.g;
  var Companion_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_datetime.$_$.e;
  var Companion_getInstance_0 = kotlin_org_jetbrains_kotlinx_kotlinx_datetime.$_$.c;
  var Companion_getInstance_1 = kotlin_org_jetbrains_kotlinx_kotlinx_datetime.$_$.f;
  var toLocalDateTime = kotlin_org_jetbrains_kotlinx_kotlinx_datetime.$_$.j;
  var _Char___init__impl__6a9atx = kotlin_kotlin.$_$.y1;
  var endsWith = kotlin_kotlin.$_$.jb;
  var protoOf = kotlin_kotlin.$_$.ia;
  var initMetadataForObject = kotlin_kotlin.$_$.n9;
  var initMetadataForCompanion = kotlin_kotlin.$_$.j9;
  var VOID = kotlin_kotlin.$_$.e;
  var LocalDateTime_init_$Create$ = kotlin_org_jetbrains_kotlinx_kotlinx_datetime.$_$.a;
  var THROW_CCE = kotlin_kotlin.$_$.md;
  var toInstant = kotlin_org_jetbrains_kotlinx_kotlinx_datetime.$_$.i;
  var Comparable = kotlin_kotlin.$_$.xc;
  var initMetadataForClass = kotlin_kotlin.$_$.i9;
  var STRING_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.g;
  var PrimitiveSerialDescriptor = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.a1;
  var KSerializer = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.j2;
  var defineProp = kotlin_kotlin.$_$.z8;
  var System_instance = kotlin_org_jetbrains_kotlinx_kotlinx_datetime.$_$.b;
  var lazy = kotlin_kotlin.$_$.ie;
  var toLong = kotlin_kotlin.$_$.ka;
  var Formats_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_datetime.$_$.d;
  var format = kotlin_org_jetbrains_kotlinx_kotlinx_datetime.$_$.h;
  var THROW_IAE = kotlin_kotlin.$_$.nd;
  var Unit_instance = kotlin_kotlin.$_$.k4;
  var Enum = kotlin_kotlin.$_$.bd;
  var Companion_getInstance_2 = kotlin_kotlin.$_$.a4;
  var hexToByteArray = kotlin_kotlin.$_$.lb;
  var toHexString = kotlin_kotlin.$_$.jc;
  var Companion_instance = kotlin_io_matthewnelson_encoding_core.$_$.c;
  var Companion_instance_0 = kotlin_io_matthewnelson_encoding_core.$_$.d;
  var encodeToByteArray = kotlin_kotlin.$_$.ib;
  var noWhenBranchMatchedException = kotlin_kotlin.$_$.je;
  var decodeToString = kotlin_kotlin.$_$.hb;
  var Base64 = kotlin_io_matthewnelson_encoding_base64.$_$.c;
  var Severity_Info_getInstance = kotlin_co_touchlab_kermit_core.$_$.f;
  var Companion_getInstance_3 = kotlin_co_touchlab_kermit.$_$.a;
  var Severity_Debug_getInstance = kotlin_co_touchlab_kermit_core.$_$.d;
  var Severity_Warn_getInstance = kotlin_co_touchlab_kermit_core.$_$.g;
  var Severity_Error_getInstance = kotlin_co_touchlab_kermit_core.$_$.e;
  var toDouble = kotlin_kotlin.$_$.ic;
  var objectCreate = kotlin_kotlin.$_$.ha;
  var Number_0 = kotlin_kotlin.$_$.id;
  var numberToLong = kotlin_kotlin.$_$.ga;
  var _ULong___init__impl__c78o9k = kotlin_kotlin.$_$.y2;
  var _UInt___init__impl__l7qpdl = kotlin_kotlin.$_$.p2;
  var _ULong___get_data__impl__fggpzb = kotlin_kotlin.$_$.z2;
  var LONG_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.f;
  var Companion_getInstance_4 = kotlin_kotlin.$_$.c4;
  //endregion
  //region block: pre-declaration
  initMetadataForObject(Static, 'Static');
  initMetadataForCompanion(Companion);
  initMetadataForClass(LocalDateTimeKMP, 'LocalDateTimeKMP', VOID, VOID, [Comparable], VOID, VOID, {0: LocalDateTimeIso8601SerializerKMP_getInstance});
  initMetadataForObject(LocalDateTimeIso8601SerializerKMP, 'LocalDateTimeIso8601SerializerKMP', VOID, VOID, [KSerializer]);
  initMetadataForObject(Static_0, 'Static');
  initMetadataForClass(DateTimeUtils, 'DateTimeUtils', DateTimeUtils);
  initMetadataForObject(DefaultDateTimeUtils, 'DefaultDateTimeUtils');
  initMetadataForClass(Encoding, 'Encoding', VOID, Enum);
  initMetadataForObject(Static_1, 'Static');
  initMetadataForClass(Logger, 'Logger', Logger);
  initMetadataForCompanion(Companion_0);
  initMetadataForClass(LongKMP, 'LongKMP', VOID, Number_0, [Number_0, Comparable], VOID, VOID, {0: LongKMPSerializer_getInstance});
  initMetadataForObject(LongKMPSerializer, 'LongKMPSerializer', VOID, VOID, [KSerializer]);
  initMetadataForObject(Uuid, 'Uuid');
  //endregion
  function kmpListOf(elements) {
    return toList(elements);
  }
  function kmpSetOf(elements) {
    return toSet(elements);
  }
  function kmpMapOf() {
    // Inline function 'kotlin.collections.mutableMapOf' call
    return LinkedHashMap_init_$Create$();
  }
  function mergeJsonElement(_this__u8e3s4, key, value) {
    var result = toMutableMap(_this__u8e3s4);
    result.t2(mapOf(new Pair(key, value)));
    return new JsonObject(result);
  }
  function Static() {
  }
  protoOf(Static).fromString = function (value) {
    var tmp;
    // Inline function 'kotlin.text.lowercase' call
    // Inline function 'kotlin.js.asDynamic' call
    var tmp$ret$1 = value.toLowerCase();
    if (endsWith(tmp$ret$1, _Char___init__impl__6a9atx(122))) {
      var instant = Companion_getInstance_0().p1l(value);
      tmp = toLocalDateTime(instant, Companion_getInstance_1().m1m(Static_getInstance_0().DEFAULT.timeZoneId));
    } else {
      tmp = Companion_getInstance().z1l(value);
    }
    var datetime = tmp;
    return new LocalDateTimeKMP(datetime.a17(), datetime.b17(), datetime.c17(), datetime.c18(), datetime.g18(), datetime.k18(), datetime.i18());
  };
  var Static_instance;
  function Static_getInstance() {
    return Static_instance;
  }
  function Companion() {
  }
  protoOf(Companion).c1q = function () {
    return LocalDateTimeIso8601SerializerKMP_getInstance();
  };
  var Companion_instance_1;
  function Companion_getInstance_5() {
    return Companion_instance_1;
  }
  function LocalDateTimeKMP(year, monthNumber, dayOfMonth, hour, minute, second, nanosecond) {
    second = second === VOID ? 0 : second;
    nanosecond = nanosecond === VOID ? 0 : nanosecond;
    this.year = year;
    this.monthNumber = monthNumber;
    this.dayOfMonth = dayOfMonth;
    this.hour = hour;
    this.minute = minute;
    this.second = second;
    this.nanosecond = nanosecond;
    this.delegate = LocalDateTime_init_$Create$(this.year, this.monthNumber, this.dayOfMonth, this.hour, this.minute, this.second, this.nanosecond);
  }
  protoOf(LocalDateTimeKMP).a17 = function () {
    return this.year;
  };
  protoOf(LocalDateTimeKMP).b17 = function () {
    return this.monthNumber;
  };
  protoOf(LocalDateTimeKMP).c17 = function () {
    return this.dayOfMonth;
  };
  protoOf(LocalDateTimeKMP).c18 = function () {
    return this.hour;
  };
  protoOf(LocalDateTimeKMP).g18 = function () {
    return this.minute;
  };
  protoOf(LocalDateTimeKMP).k18 = function () {
    return this.second;
  };
  protoOf(LocalDateTimeKMP).i18 = function () {
    return this.nanosecond;
  };
  protoOf(LocalDateTimeKMP).b2h = function () {
    return this.delegate;
  };
  protoOf(LocalDateTimeKMP).c2h = function (other) {
    return Static_instance.fromString(this.toString()).c2h(Static_instance.fromString(other.toString()));
  };
  protoOf(LocalDateTimeKMP).d = function (other) {
    return this.c2h(other instanceof LocalDateTimeKMP ? other : THROW_CCE());
  };
  protoOf(LocalDateTimeKMP).toString = function () {
    return this.delegate.toString();
  };
  protoOf(LocalDateTimeKMP).d2h = function (timeZoneId) {
    return fromLong(toInstant(this.delegate, Static_getInstance_0().DEFAULT.e2h(timeZoneId)).r1a());
  };
  protoOf(LocalDateTimeKMP).toEpochSeconds = function (timeZoneId, $super) {
    timeZoneId = timeZoneId === VOID ? Companion_getInstance_1().l1m_1.p1m() : timeZoneId;
    return this.d2h(timeZoneId);
  };
  function LocalDateTimeIso8601SerializerKMP() {
    LocalDateTimeIso8601SerializerKMP_instance = this;
    this.f2h_1 = PrimitiveSerialDescriptor('com.sphereon.kmp.datetime.LocalDateTime', STRING_getInstance());
  }
  protoOf(LocalDateTimeIso8601SerializerKMP).hl = function () {
    return this.f2h_1;
  };
  protoOf(LocalDateTimeIso8601SerializerKMP).vl = function (decoder) {
    var parsed = Companion_getInstance().z1l(decoder.fo());
    return new LocalDateTimeKMP(parsed.a17(), parsed.b17(), parsed.c17(), parsed.c18(), parsed.g18(), parsed.k18(), parsed.i18());
  };
  protoOf(LocalDateTimeIso8601SerializerKMP).g2h = function (encoder, value) {
    encoder.mp(value.toString());
  };
  protoOf(LocalDateTimeIso8601SerializerKMP).ul = function (encoder, value) {
    return this.g2h(encoder, value instanceof LocalDateTimeKMP ? value : THROW_CCE());
  };
  var LocalDateTimeIso8601SerializerKMP_instance;
  function LocalDateTimeIso8601SerializerKMP_getInstance() {
    if (LocalDateTimeIso8601SerializerKMP_instance == null)
      new LocalDateTimeIso8601SerializerKMP();
    return LocalDateTimeIso8601SerializerKMP_instance;
  }
  function Static_0() {
    Static_instance_0 = this;
    this.DEFAULT = new DateTimeUtils();
  }
  protoOf(Static_0).h2h = function () {
    return this.DEFAULT;
  };
  var Static_instance_0;
  function Static_getInstance_0() {
    if (Static_instance_0 == null)
      new Static_0();
    return Static_instance_0;
  }
  function now($this) {
    return $this.clock.r16();
  }
  function DateTimeUtils$defaultTimeZone$lambda(this$0) {
    return function () {
      return Companion_getInstance_1().m1m(this$0.timeZoneId);
    };
  }
  function DateTimeUtils(clock, timeZoneId) {
    clock = clock === VOID ? System_instance : clock;
    timeZoneId = timeZoneId === VOID ? Companion_getInstance_1().c1n().p1m() : timeZoneId;
    this.clock = clock;
    this.timeZoneId = timeZoneId;
    var tmp = this;
    tmp.a2h_1 = lazy(DateTimeUtils$defaultTimeZone$lambda(this));
  }
  protoOf(DateTimeUtils).i2h = function (_set____db54di) {
    this.clock = _set____db54di;
  };
  protoOf(DateTimeUtils).j2h = function () {
    return this.clock;
  };
  protoOf(DateTimeUtils).k2h = function (_set____db54di) {
    this.timeZoneId = _set____db54di;
  };
  protoOf(DateTimeUtils).l2h = function () {
    return this.timeZoneId;
  };
  protoOf(DateTimeUtils).epochSeconds = function () {
    return now(this).r1a().i1();
  };
  protoOf(DateTimeUtils).m2h = function (epochSeconds) {
    return this.dateTime(Companion_getInstance_1().l1m_1.p1m(), epochSeconds);
  };
  protoOf(DateTimeUtils).dateTimeUTC = function (epochSeconds, $super) {
    epochSeconds = epochSeconds === VOID ? this.epochSeconds() : epochSeconds;
    return this.m2h(epochSeconds);
  };
  protoOf(DateTimeUtils).n2h = function (epochSeconds) {
    return this.dateTime(this.timeZoneId, epochSeconds);
  };
  protoOf(DateTimeUtils).dateTimeLocal = function (epochSeconds, $super) {
    epochSeconds = epochSeconds === VOID ? this.epochSeconds() : epochSeconds;
    return this.n2h(epochSeconds);
  };
  protoOf(DateTimeUtils).o2h = function (timeZoneId, epochSeconds) {
    var tmp = Companion_getInstance_0();
    var tmp1_elvis_lhs = epochSeconds == null ? null : toLong(epochSeconds);
    return toKMP(toLocalDateTime(tmp.u1m(tmp1_elvis_lhs == null ? toLong(this.epochSeconds()) : tmp1_elvis_lhs), this.e2h(timeZoneId)));
  };
  protoOf(DateTimeUtils).dateTime = function (timeZoneId, epochSeconds, $super) {
    timeZoneId = timeZoneId === VOID ? null : timeZoneId;
    epochSeconds = epochSeconds === VOID ? this.epochSeconds() : epochSeconds;
    return this.o2h(timeZoneId, epochSeconds);
  };
  protoOf(DateTimeUtils).p2h = function (epochSeconds) {
    return this.dateISO(this.timeZoneId, epochSeconds);
  };
  protoOf(DateTimeUtils).dateLocalISO = function (epochSeconds, $super) {
    epochSeconds = epochSeconds === VOID ? this.epochSeconds() : epochSeconds;
    return this.p2h(epochSeconds);
  };
  protoOf(DateTimeUtils).q2h = function (timeZoneId, epochSeconds) {
    return format(toKotlin(this.dateTime(timeZoneId == null ? this.timeZoneId : timeZoneId, epochSeconds)).q1e(), Formats_getInstance().x16());
  };
  protoOf(DateTimeUtils).dateISO = function (timeZoneId, epochSeconds, $super) {
    epochSeconds = epochSeconds === VOID ? this.epochSeconds() : epochSeconds;
    return this.q2h(timeZoneId, epochSeconds);
  };
  protoOf(DateTimeUtils).e2h = function (timeZoneId) {
    var tmp;
    if (timeZoneId == null) {
      tmp = null;
    } else {
      // Inline function 'kotlin.let' call
      // Inline function 'kotlin.contracts.contract' call
      // Inline function 'com.sphereon.kmp.DateTimeUtils.timeZone.<anonymous>' call
      tmp = Companion_getInstance_1().m1m(timeZoneId);
    }
    var tmp1_elvis_lhs = tmp;
    return tmp1_elvis_lhs == null ? this.a2h_1.d2() : tmp1_elvis_lhs;
  };
  function DefaultDateTimeUtils() {
    DefaultDateTimeUtils_instance = this;
    this.INSTANCE = new DateTimeUtils();
  }
  protoOf(DefaultDateTimeUtils).r2h = function () {
    return this.INSTANCE;
  };
  var DefaultDateTimeUtils_instance;
  function DefaultDateTimeUtils_getInstance() {
    if (DefaultDateTimeUtils_instance == null)
      new DefaultDateTimeUtils();
    return DefaultDateTimeUtils_instance;
  }
  function toKMP(_this__u8e3s4) {
    return Static_instance.fromString(_this__u8e3s4.toString());
  }
  function toKotlin(_this__u8e3s4) {
    return Companion_getInstance().z1l(_this__u8e3s4.toString());
  }
  function getDateTime(dateTimeUtils) {
    dateTimeUtils = dateTimeUtils === VOID ? null : dateTimeUtils;
    return dateTimeUtils == null ? DefaultDateTimeUtils_getInstance().INSTANCE : dateTimeUtils;
  }
  function toLocalDateTimeKMP(_this__u8e3s4, dateTimeUtils) {
    dateTimeUtils = dateTimeUtils === VOID ? null : dateTimeUtils;
    return getDateTime(dateTimeUtils).dateTimeLocal(_this__u8e3s4.i1());
  }
  function get_base64UrlConf() {
    _init_properties_Encoding_kt__jty2h9();
    return base64UrlConf;
  }
  var base64UrlConf;
  function get_base64Conf() {
    _init_properties_Encoding_kt__jty2h9();
    return base64Conf;
  }
  var base64Conf;
  var Encoding_BASE64_instance;
  var Encoding_BASE64URL_instance;
  var Encoding_HEX_instance;
  var Encoding_UTF8_instance;
  function values() {
    return [Encoding_BASE64_getInstance(), Encoding_BASE64URL_getInstance(), Encoding_HEX_getInstance(), Encoding_UTF8_getInstance()];
  }
  function valueOf(value) {
    switch (value) {
      case 'BASE64':
        return Encoding_BASE64_getInstance();
      case 'BASE64URL':
        return Encoding_BASE64URL_getInstance();
      case 'HEX':
        return Encoding_HEX_getInstance();
      case 'UTF8':
        return Encoding_UTF8_getInstance();
      default:
        Encoding_initEntries();
        THROW_IAE('No enum constant value.');
        break;
    }
  }
  var Encoding_entriesInitialized;
  function Encoding_initEntries() {
    if (Encoding_entriesInitialized)
      return Unit_instance;
    Encoding_entriesInitialized = true;
    Encoding_BASE64_instance = new Encoding('BASE64', 0);
    Encoding_BASE64URL_instance = new Encoding('BASE64URL', 1);
    Encoding_HEX_instance = new Encoding('HEX', 2);
    Encoding_UTF8_instance = new Encoding('UTF8', 3);
  }
  function Encoding(name, ordinal) {
    Enum.call(this, name, ordinal);
  }
  function decodeFromHex(_this__u8e3s4) {
    _init_properties_Encoding_kt__jty2h9();
    return hexToByteArray(_this__u8e3s4, Companion_getInstance_2().sg_1);
  }
  function encodeToHex(_this__u8e3s4) {
    _init_properties_Encoding_kt__jty2h9();
    return toHexString(_this__u8e3s4, Companion_getInstance_2().sg_1);
  }
  function decodeFromBase64Url(_this__u8e3s4) {
    _init_properties_Encoding_kt__jty2h9();
    return Companion_instance.t2d(_this__u8e3s4, get_base64UrlConf());
  }
  function encodeToBase64Url(_this__u8e3s4) {
    _init_properties_Encoding_kt__jty2h9();
    return Companion_instance_0.e2e(_this__u8e3s4, get_base64UrlConf());
  }
  function encodeToBase64(_this__u8e3s4, urlSafe) {
    urlSafe = urlSafe === VOID ? false : urlSafe;
    _init_properties_Encoding_kt__jty2h9();
    return Companion_instance_0.e2e(_this__u8e3s4, urlSafe ? get_base64UrlConf() : get_base64Conf());
  }
  function decodeFromBase64(_this__u8e3s4, urlSafe) {
    urlSafe = urlSafe === VOID ? false : urlSafe;
    _init_properties_Encoding_kt__jty2h9();
    return Companion_instance.t2d(_this__u8e3s4, urlSafe ? get_base64UrlConf() : get_base64Conf());
  }
  function decodeFrom(_this__u8e3s4, encoding) {
    _init_properties_Encoding_kt__jty2h9();
    var tmp;
    switch (encoding.v2_1) {
      case 1:
        tmp = decodeFromBase64Url(_this__u8e3s4);
        break;
      case 0:
        tmp = decodeFromBase64(_this__u8e3s4, false);
        break;
      case 2:
        tmp = decodeFromHex(_this__u8e3s4);
        break;
      case 3:
        tmp = encodeToByteArray(_this__u8e3s4);
        break;
      default:
        noWhenBranchMatchedException();
        break;
    }
    return tmp;
  }
  function encodeTo(_this__u8e3s4, encoding) {
    _init_properties_Encoding_kt__jty2h9();
    var tmp;
    switch (encoding.v2_1) {
      case 1:
        tmp = encodeToBase64Url(_this__u8e3s4);
        break;
      case 0:
        tmp = encodeToBase64(_this__u8e3s4, false);
        break;
      case 2:
        tmp = encodeToHex(_this__u8e3s4);
        break;
      case 3:
        tmp = decodeToString(_this__u8e3s4);
        break;
      default:
        noWhenBranchMatchedException();
        break;
    }
    return tmp;
  }
  function base64UrlConf$lambda($this$Base64) {
    _init_properties_Encoding_kt__jty2h9();
    $this$Base64.k2f_1 = true;
    $this$Base64.l2f_1 = 0;
    $this$Base64.m2f_1 = true;
    $this$Base64.n2f_1 = false;
    return Unit_instance;
  }
  function base64Conf$lambda($this$Base64) {
    _init_properties_Encoding_kt__jty2h9();
    $this$Base64.k2f_1 = true;
    $this$Base64.l2f_1 = 0;
    $this$Base64.m2f_1 = false;
    $this$Base64.n2f_1 = false;
    return Unit_instance;
  }
  function Encoding_BASE64_getInstance() {
    Encoding_initEntries();
    return Encoding_BASE64_instance;
  }
  function Encoding_BASE64URL_getInstance() {
    Encoding_initEntries();
    return Encoding_BASE64URL_instance;
  }
  function Encoding_HEX_getInstance() {
    Encoding_initEntries();
    return Encoding_HEX_instance;
  }
  function Encoding_UTF8_getInstance() {
    Encoding_initEntries();
    return Encoding_UTF8_instance;
  }
  var properties_initialized_Encoding_kt_vg5col;
  function _init_properties_Encoding_kt__jty2h9() {
    if (!properties_initialized_Encoding_kt_vg5col) {
      properties_initialized_Encoding_kt_vg5col = true;
      base64UrlConf = Base64(base64UrlConf$lambda);
      base64Conf = Base64(base64Conf$lambda);
    }
  }
  function Static_1() {
  }
  protoOf(Static_1).u2h = function (tag, severity) {
    // Inline function 'kotlin.also' call
    var this_0 = new Logger(tag);
    // Inline function 'kotlin.contracts.contract' call
    // Inline function 'com.sphereon.kmp.Static.tag.<anonymous>' call
    this_0.z2c(severity);
    return this_0;
  };
  protoOf(Static_1).w2h = function (tag, severity, $super) {
    tag = tag === VOID ? '' : tag;
    severity = severity === VOID ? Severity_Info_getInstance() : severity;
    return $super === VOID ? this.u2h(tag, severity) : $super.u2h.call(this, tag, severity);
  };
  var Static_instance_1;
  function Static_getInstance_1() {
    return Static_instance_1;
  }
  function Logger(tag) {
    tag = tag === VOID ? '' : tag;
    this.v2h_1 = tag;
  }
  protoOf(Logger).x2h = function (message, tag) {
    // Inline function 'co.touchlab.kermit.Logger.d' call
    // Inline function 'co.touchlab.kermit.BaseLogger.log' call
    var this_0 = Companion_getInstance_3();
    var severity = Severity_Debug_getInstance();
    if (this_0.d2c().w2c().y2(severity) <= 0) {
      this_0.f2c(severity, tag, null, message);
    }
  };
  protoOf(Logger).y2h = function (message, tag, $super) {
    tag = tag === VOID ? this.v2h_1 : tag;
    var tmp;
    if ($super === VOID) {
      this.x2h(message, tag);
      tmp = Unit_instance;
    } else {
      tmp = $super.x2h.call(this, message, tag);
    }
    return tmp;
  };
  protoOf(Logger).z2h = function (message, tag) {
    // Inline function 'co.touchlab.kermit.Logger.i' call
    // Inline function 'co.touchlab.kermit.BaseLogger.log' call
    var this_0 = Companion_getInstance_3();
    var severity = Severity_Info_getInstance();
    if (this_0.d2c().w2c().y2(severity) <= 0) {
      this_0.f2c(severity, tag, null, message);
    }
  };
  protoOf(Logger).a2i = function (message, tag, $super) {
    tag = tag === VOID ? this.v2h_1 : tag;
    var tmp;
    if ($super === VOID) {
      this.z2h(message, tag);
      tmp = Unit_instance;
    } else {
      tmp = $super.z2h.call(this, message, tag);
    }
    return tmp;
  };
  protoOf(Logger).b2i = function (message, tag) {
    // Inline function 'co.touchlab.kermit.Logger.w' call
    // Inline function 'co.touchlab.kermit.BaseLogger.log' call
    var this_0 = Companion_getInstance_3();
    var severity = Severity_Warn_getInstance();
    if (this_0.d2c().w2c().y2(severity) <= 0) {
      this_0.f2c(severity, tag, null, message);
    }
  };
  protoOf(Logger).c2i = function (message, tag, $super) {
    tag = tag === VOID ? this.v2h_1 : tag;
    var tmp;
    if ($super === VOID) {
      this.b2i(message, tag);
      tmp = Unit_instance;
    } else {
      tmp = $super.b2i.call(this, message, tag);
    }
    return tmp;
  };
  protoOf(Logger).d2i = function (message, throwable, tag) {
    // Inline function 'co.touchlab.kermit.Logger.e' call
    // Inline function 'co.touchlab.kermit.BaseLogger.log' call
    var this_0 = Companion_getInstance_3();
    var severity = Severity_Error_getInstance();
    if (this_0.d2c().w2c().y2(severity) <= 0) {
      this_0.f2c(severity, tag, throwable, message);
    }
  };
  protoOf(Logger).e2i = function (message, throwable, tag, $super) {
    throwable = throwable === VOID ? null : throwable;
    tag = tag === VOID ? this.v2h_1 : tag;
    var tmp;
    if ($super === VOID) {
      this.d2i(message, throwable, tag);
      tmp = Unit_instance;
    } else {
      tmp = $super.d2i.call(this, message, throwable, tag);
    }
    return tmp;
  };
  protoOf(Logger).z2c = function (severity) {
    return Companion_getInstance_3().z2c(severity);
  };
  function LongKMP_init_$Init$(value, $this) {
    LongKMP.call($this, toDouble(value));
    return $this;
  }
  function fromString(value) {
    return LongKMP_init_$Init$(value, objectCreate(protoOf(LongKMP)));
  }
  function LongKMP_init_$Init$_0(value, $this) {
    LongKMP.call($this, value.u3());
    return $this;
  }
  function fromLong(value) {
    return LongKMP_init_$Init$_0(value, objectCreate(protoOf(LongKMP)));
  }
  function LongKMP_init_$Init$_1(value, $this) {
    LongKMP.call($this, value);
    return $this;
  }
  function fromNumber(value) {
    return LongKMP_init_$Init$_1(value, objectCreate(protoOf(LongKMP)));
  }
  function Companion_0() {
  }
  protoOf(Companion_0).c1q = function () {
    return LongKMPSerializer_getInstance();
  };
  var Companion_instance_2;
  function Companion_getInstance_6() {
    return Companion_instance_2;
  }
  function LongKMP(value) {
    Number_0.call(this);
    this.f2i_1 = numberToLong(value);
  }
  protoOf(LongKMP).r3 = function () {
    return this.f2i_1.r3();
  };
  protoOf(LongKMP).u3 = function () {
    return this.f2i_1.u3();
  };
  protoOf(LongKMP).t3 = function () {
    return this.f2i_1.t3();
  };
  protoOf(LongKMP).i1 = function () {
    return this.f2i_1.i1();
  };
  protoOf(LongKMP).g2i = function () {
    return this.f2i_1;
  };
  protoOf(LongKMP).toULong = function () {
    // Inline function 'kotlin.toULong' call
    var this_0 = this.f2i_1;
    return _ULong___init__impl__c78o9k(this_0);
  };
  protoOf(LongKMP).toUInt = function () {
    // Inline function 'kotlin.toUInt' call
    var this_0 = this.f2i_1;
    return _UInt___init__impl__l7qpdl(this_0.i1());
  };
  protoOf(LongKMP).s3 = function () {
    return this.f2i_1.s3();
  };
  protoOf(LongKMP).toString = function () {
    return this.f2i_1.toString();
  };
  protoOf(LongKMP).h2i = function (other) {
    return this.f2i_1.g1(other.f2i_1);
  };
  protoOf(LongKMP).d = function (other) {
    return this.h2i(other instanceof LongKMP ? other : THROW_CCE());
  };
  protoOf(LongKMP).equals = function (other) {
    if (this === other)
      return true;
    if (!(other instanceof LongKMP))
      return false;
    if (!this.f2i_1.equals(other.f2i_1))
      return false;
    return true;
  };
  protoOf(LongKMP).hashCode = function () {
    return this.f2i_1.hashCode();
  };
  function numberToKmpLong(_this__u8e3s4) {
    return toKmpLong(_this__u8e3s4);
  }
  function ulongToKmpLong(_this__u8e3s4) {
    // Inline function 'kotlin.ULong.toLong' call
    var tmp$ret$0 = _ULong___get_data__impl__fggpzb(_this__u8e3s4);
    return fromLong(tmp$ret$0);
  }
  function byteToKmpLong(_this__u8e3s4) {
    return fromLong(toLong(_this__u8e3s4));
  }
  function stringToKmpLong(_this__u8e3s4) {
    return fromString(_this__u8e3s4);
  }
  function LongKMPSerializer() {
    LongKMPSerializer_instance = this;
    this.i2i_1 = PrimitiveSerialDescriptor('LongKMP', LONG_getInstance());
  }
  protoOf(LongKMPSerializer).hl = function () {
    return this.i2i_1;
  };
  protoOf(LongKMPSerializer).j2i = function (encoder, value) {
    encoder.ip(value.g2i());
  };
  protoOf(LongKMPSerializer).ul = function (encoder, value) {
    return this.j2i(encoder, value instanceof LongKMP ? value : THROW_CCE());
  };
  protoOf(LongKMPSerializer).vl = function (decoder) {
    return fromLong(decoder.bo());
  };
  var LongKMPSerializer_instance;
  function LongKMPSerializer_getInstance() {
    if (LongKMPSerializer_instance == null)
      new LongKMPSerializer();
    return LongKMPSerializer_instance;
  }
  function toKmpLong(_this__u8e3s4) {
    return fromLong(numberToLong(_this__u8e3s4));
  }
  function Uuid() {
  }
  protoOf(Uuid).v4 = function () {
    return Companion_getInstance_4().fj();
  };
  protoOf(Uuid).v4String = function () {
    return this.v4().toString();
  };
  var Uuid_instance;
  function Uuid_getInstance() {
    return Uuid_instance;
  }
  //region block: post-declaration
  defineProp(protoOf(Encoding), 'name', protoOf(Encoding).w2);
  defineProp(protoOf(Encoding), 'ordinal', protoOf(Encoding).x2);
  //endregion
  //region block: init
  Static_instance = new Static();
  Companion_instance_1 = new Companion();
  Static_instance_1 = new Static_1();
  Companion_instance_2 = new Companion_0();
  Uuid_instance = new Uuid();
  //endregion
  //region block: exports
  function $jsExportAll$(_) {
    var $com = _.com || (_.com = {});
    var $com$sphereon = $com.sphereon || ($com.sphereon = {});
    var $com$sphereon$kmp = $com$sphereon.kmp || ($com$sphereon.kmp = {});
    $com$sphereon$kmp.kmpListOf = kmpListOf;
    $com$sphereon$kmp.kmpSetOf = kmpSetOf;
    $com$sphereon$kmp.kmpMapOf = kmpMapOf;
    var $com = _.com || (_.com = {});
    var $com$sphereon = $com.sphereon || ($com.sphereon = {});
    var $com$sphereon$kmp = $com$sphereon.kmp || ($com$sphereon.kmp = {});
    $com$sphereon$kmp.LocalDateTimeKMP = LocalDateTimeKMP;
    defineProp($com$sphereon$kmp.LocalDateTimeKMP, 'Static', Static_getInstance);
    defineProp($com$sphereon$kmp.LocalDateTimeKMP, 'Companion', Companion_getInstance_5);
    var $com = _.com || (_.com = {});
    var $com$sphereon = $com.sphereon || ($com.sphereon = {});
    var $com$sphereon$kmp = $com$sphereon.kmp || ($com$sphereon.kmp = {});
    $com$sphereon$kmp.DateTimeUtils = DateTimeUtils;
    defineProp($com$sphereon$kmp.DateTimeUtils, 'Static', Static_getInstance_0);
    defineProp($com$sphereon$kmp, 'DefaultDateTimeUtils', DefaultDateTimeUtils_getInstance);
    var $com = _.com || (_.com = {});
    var $com$sphereon = $com.sphereon || ($com.sphereon = {});
    var $com$sphereon$kmp = $com$sphereon.kmp || ($com$sphereon.kmp = {});
    $com$sphereon$kmp.Encoding = Encoding;
    $com$sphereon$kmp.Encoding.values = values;
    $com$sphereon$kmp.Encoding.valueOf = valueOf;
    defineProp($com$sphereon$kmp.Encoding, 'BASE64', Encoding_BASE64_getInstance);
    defineProp($com$sphereon$kmp.Encoding, 'BASE64URL', Encoding_BASE64URL_getInstance);
    defineProp($com$sphereon$kmp.Encoding, 'HEX', Encoding_HEX_getInstance);
    defineProp($com$sphereon$kmp.Encoding, 'UTF8', Encoding_UTF8_getInstance);
    $com$sphereon$kmp.decodeFromHex = decodeFromHex;
    $com$sphereon$kmp.encodeToHex = encodeToHex;
    $com$sphereon$kmp.decodeFromBase64Url = decodeFromBase64Url;
    $com$sphereon$kmp.encodeToBase64Url = encodeToBase64Url;
    $com$sphereon$kmp.encodeToBase64 = encodeToBase64;
    $com$sphereon$kmp.decodeFromBase64 = decodeFromBase64;
    $com$sphereon$kmp.decodeFrom = decodeFrom;
    $com$sphereon$kmp.encodeTo = encodeTo;
    var $com = _.com || (_.com = {});
    var $com$sphereon = $com.sphereon || ($com.sphereon = {});
    var $com$sphereon$kmp = $com$sphereon.kmp || ($com$sphereon.kmp = {});
    $com$sphereon$kmp.LongKMP = LongKMP;
    $com$sphereon$kmp.LongKMP.fromString = fromString;
    $com$sphereon$kmp.LongKMP.fromLong = fromLong;
    $com$sphereon$kmp.LongKMP.fromNumber = fromNumber;
    defineProp($com$sphereon$kmp.LongKMP, 'Companion', Companion_getInstance_6);
    $com$sphereon$kmp.numberToKmpLong = numberToKmpLong;
    $com$sphereon$kmp.ulongToKmpLong = ulongToKmpLong;
    $com$sphereon$kmp.byteToKmpLong = byteToKmpLong;
    $com$sphereon$kmp.stringToKmpLong = stringToKmpLong;
    var $com = _.com || (_.com = {});
    var $com$sphereon = $com.sphereon || ($com.sphereon = {});
    var $com$sphereon$kmp = $com$sphereon.kmp || ($com$sphereon.kmp = {});
    defineProp($com$sphereon$kmp, 'Uuid', Uuid_getInstance);
  }
  $jsExportAll$(_);
  _.$jsExportAll$ = $jsExportAll$;
  _.$_$ = _.$_$ || {};
  _.$_$.a = LongKMP;
  _.$_$.b = decodeFromHex;
  _.$_$.c = decodeFrom;
  _.$_$.d = encodeToBase64Url;
  _.$_$.e = encodeTo;
  _.$_$.f = getDateTime;
  _.$_$.g = mergeJsonElement;
  _.$_$.h = numberToKmpLong;
  _.$_$.i = stringToKmpLong;
  _.$_$.j = toKmpLong;
  _.$_$.k = toKotlin;
  _.$_$.l = toLocalDateTimeKMP;
  _.$_$.m = ulongToKmpLong;
  _.$_$.n = Encoding_BASE64URL_getInstance;
  _.$_$.o = Encoding_BASE64_getInstance;
  _.$_$.p = Encoding_HEX_getInstance;
  _.$_$.q = Encoding_UTF8_getInstance;
  _.$_$.r = fromNumber;
  _.$_$.s = fromLong;
  _.$_$.t = Static_getInstance_0;
  _.$_$.u = DefaultDateTimeUtils_getInstance;
  _.$_$.v = Static_instance;
  _.$_$.w = Static_instance_1;
  _.$_$.x = LongKMPSerializer_getInstance;
  _.$_$.y = Uuid_instance;
  //endregion
  return _;
}));
