// @hyperledger/anoncreds-nodejs/build/NodeJSAnoncreds.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeJSAnoncreds = void 0;
const node_util_1 = require("node:util");
const anoncreds_shared_1 = require("@hyperledger/anoncreds-shared");
const ffi_1 = require("./ffi");
const library_1 = require("./library");
function handleReturnPointer(returnValue) {
    const value = Array.isArray(returnValue) ? returnValue[0] : returnValue;
    if (value === null || value === undefined) {
        throw anoncreds_shared_1.AnoncredsError.customError({ message: 'Unexpected null pointer' });
    }
    return value;
}
class NodeJSAnoncreds {
    get nativeAnoncreds() {
        return (0, library_1.getNativeAnonCreds)();
    }
    /**
     * Fetch the error from the native library and throw it as a JS error
     *
     * NOTE:
     * Checks whether the error code of the returned error matches the error code that was passed to the function.
     * If it doesn't, we throw an error with the original errorCode, and a custom message explaining we weren't able
     * to retrieve the error message from the native library. This should however not break functionality as long as
     * error codes are used rather than error messages for error handling.
     *
     */
    getAnoncredsError(errorCode) {
        const error = this.getCurrentError();
        if (error.code !== errorCode) {
            return new anoncreds_shared_1.AnoncredsError({
                code: errorCode,
                message: 'Error details have already been overwritten on the native side, unable to retrieve error message for the error',
            });
        }
        return new anoncreds_shared_1.AnoncredsError(error);
    }
    getCurrentError() {
        const error = (0, ffi_1.allocateStringBuffer)();
        this.nativeAnoncreds.anoncreds_get_current_error(error);
        const serializedError = handleReturnPointer(error);
        return JSON.parse(serializedError);
    }
    handleError(errorCode) {
        if (errorCode === 0)
            return;
        throw this.getAnoncredsError(errorCode);
    }
    generateNonce() {
        const ret = (0, ffi_1.allocateStringBuffer)();
        const errorCode = this.nativeAnoncreds.anoncreds_generate_nonce(ret);
        this.handleError(errorCode);
        return handleReturnPointer(ret);
    }
    createSchema(options) {
        const { name, version, issuerId, attributeNames } = (0, ffi_1.serializeArguments)(options);
        const ret = (0, ffi_1.allocatePointer)();
        const errorCode = this.nativeAnoncreds.anoncreds_create_schema(name, version, issuerId, attributeNames, ret);
        this.handleError(errorCode);
        return new anoncreds_shared_1.ObjectHandle(handleReturnPointer(ret));
    }
    revocationRegistryDefinitionGetAttribute(options) {
        const { objectHandle, name } = (0, ffi_1.serializeArguments)(options);
        const ret = (0, ffi_1.allocateStringBuffer)();
        const errorCode = this.nativeAnoncreds.anoncreds_revocation_registry_definition_get_attribute(objectHandle, name, ret);
        this.handleError(errorCode);
        return handleReturnPointer(ret);
    }
    credentialGetAttribute(options) {
        var _a;
        const { objectHandle, name } = (0, ffi_1.serializeArguments)(options);
        const ret = (0, ffi_1.allocateStringBuffer)();
        const errorCode = this.nativeAnoncreds.anoncreds_credential_get_attribute(objectHandle, name, ret);
        this.handleError(errorCode);
        const value = Array.isArray(ret) ? ret[0] : ret;
        return (_a = value) !== null && _a !== void 0 ? _a : null;
    }
    createCredentialDefinition(options) {
        const { schemaId, issuerId, schema, tag, signatureType, supportRevocation } = (0, ffi_1.serializeArguments)(options);
        const credentialDefinitionPtr = (0, ffi_1.allocatePointer)();
        const credentialDefinitionPrivatePtr = (0, ffi_1.allocatePointer)();
        const keyCorrectnessProofPtr = (0, ffi_1.allocatePointer)();
        const errorCode = this.nativeAnoncreds.anoncreds_create_credential_definition(schemaId, schema, tag, issuerId, signatureType, supportRevocation, credentialDefinitionPtr, credentialDefinitionPrivatePtr, keyCorrectnessProofPtr);
        this.handleError(errorCode);
        return {
            credentialDefinition: new anoncreds_shared_1.ObjectHandle(handleReturnPointer(credentialDefinitionPtr)),
            credentialDefinitionPrivate: new anoncreds_shared_1.ObjectHandle(handleReturnPointer(credentialDefinitionPrivatePtr)),
            keyCorrectnessProof: new anoncreds_shared_1.ObjectHandle(handleReturnPointer(keyCorrectnessProofPtr)),
        };
    }
    createCredential(options) {
        const { credentialDefinition, credentialDefinitionPrivate, credentialOffer, credentialRequest } = (0, ffi_1.serializeArguments)(options);
        const attributeNames = {
            count: Object.keys(options.attributeRawValues).length,
            data: Object.keys(options.attributeRawValues),
        };
        const attributeRawValues = {
            count: Object.keys(options.attributeRawValues).length,
            data: Object.values(options.attributeRawValues),
        };
        const attributeEncodedValues = options.attributeEncodedValues
            ? {
                count: Object.keys(options.attributeEncodedValues).length,
                data: Object.values(options.attributeEncodedValues),
            }
            : {};
        let revocationConfiguration = null;
        if (options.revocationConfiguration) {
            const { revocationStatusList, revocationRegistryDefinitionPrivate, revocationRegistryDefinition, registryIndex } = (0, ffi_1.serializeArguments)(options.revocationConfiguration);
            revocationConfiguration = {
                reg_def: revocationRegistryDefinition,
                reg_def_private: revocationRegistryDefinitionPrivate,
                status_list: revocationStatusList,
                reg_idx: registryIndex,
            };
        }
        const credentialPtr = (0, ffi_1.allocatePointer)();
        const errorCode = this.nativeAnoncreds.anoncreds_create_credential(credentialDefinition, credentialDefinitionPrivate, credentialOffer, credentialRequest, attributeNames, attributeRawValues, attributeEncodedValues, revocationConfiguration, credentialPtr);
        this.handleError(errorCode);
        return new anoncreds_shared_1.ObjectHandle(handleReturnPointer(credentialPtr));
    }
    encodeCredentialAttributes(options) {
        const { attributeRawValues } = (0, ffi_1.serializeArguments)(options);
        const ret = (0, ffi_1.allocateStringBuffer)();
        const errorCode = this.nativeAnoncreds.anoncreds_encode_credential_attributes(attributeRawValues, ret);
        this.handleError(errorCode);
        const result = handleReturnPointer(ret);
        return result.split(',');
    }
    processCredential(options) {
        var _a, _b;
        const { credential, credentialRequestMetadata, linkSecret, credentialDefinition } = (0, ffi_1.serializeArguments)(options);
        const ret = (0, ffi_1.allocatePointer)();
        const errorCode = this.nativeAnoncreds.anoncreds_process_credential(credential, credentialRequestMetadata, linkSecret, credentialDefinition, (_b = (_a = options.revocationRegistryDefinition) === null || _a === void 0 ? void 0 : _a.handle) !== null && _b !== void 0 ? _b : 0, ret);
        this.handleError(errorCode);
        return new anoncreds_shared_1.ObjectHandle(handleReturnPointer(ret));
    }
    createCredentialOffer(options) {
        const { schemaId, credentialDefinitionId, keyCorrectnessProof } = (0, ffi_1.serializeArguments)(options);
        const ret = (0, ffi_1.allocatePointer)();
        const errorCode = this.nativeAnoncreds.anoncreds_create_credential_offer(schemaId, credentialDefinitionId, keyCorrectnessProof, ret);
        this.handleError(errorCode);
        return new anoncreds_shared_1.ObjectHandle(handleReturnPointer(ret));
    }
    createCredentialRequest(options) {
        const { entropy, proverDid, credentialDefinition, linkSecret, linkSecretId, credentialOffer } = (0, ffi_1.serializeArguments)(options);
        const credentialRequestPtr = (0, ffi_1.allocatePointer)();
        const credentialRequestMetadataPtr = (0, ffi_1.allocatePointer)();
        const errorCode = this.nativeAnoncreds.anoncreds_create_credential_request(entropy, proverDid, credentialDefinition, linkSecret, linkSecretId, credentialOffer, credentialRequestPtr, credentialRequestMetadataPtr);
        this.handleError(errorCode);
        return {
            credentialRequest: new anoncreds_shared_1.ObjectHandle(handleReturnPointer(credentialRequestPtr)),
            credentialRequestMetadata: new anoncreds_shared_1.ObjectHandle(handleReturnPointer(credentialRequestMetadataPtr)),
        };
    }
    createLinkSecret() {
        const ret = (0, ffi_1.allocateStringBuffer)();
        const errorCode = this.nativeAnoncreds.anoncreds_create_link_secret(ret);
        this.handleError(errorCode);
        return handleReturnPointer(ret);
    }
    createPresentation(options) {
        const { presentationRequest, linkSecret } = (0, ffi_1.serializeArguments)(options);
        const selfAttestNames = {
            count: Object.keys(options.selfAttest).length,
            data: Object.keys(options.selfAttest),
        };
        const selfAttestValues = {
            count: Object.values(options.selfAttest).length,
            data: Object.values(options.selfAttest),
        };
        const credentialEntryList = this.convertCredentialList(options.credentials);
        const credentialProveList = this.convertCredentialProves(options.credentialsProve);
        const { schemaIds, schemas } = this.convertSchemas(options.schemas);
        const { credentialDefinitionIds, credentialDefinitions } = this.convertCredDefs(options.credentialDefinitions);
        const ret = (0, ffi_1.allocatePointer)();
        const errorCode = this.nativeAnoncreds.anoncreds_create_presentation(presentationRequest, credentialEntryList, credentialProveList, selfAttestNames, selfAttestValues, linkSecret, schemas, schemaIds, credentialDefinitions, credentialDefinitionIds, ret);
        this.handleError(errorCode);
        return new anoncreds_shared_1.ObjectHandle(handleReturnPointer(ret));
    }
    verifyPresentation(options) {
        const { presentation, presentationRequest, schemas, credentialDefinitions, revocationRegistryDefinitions, revocationStatusLists, revocationRegistryDefinitionIds, schemaIds, credentialDefinitionIds, } = (0, ffi_1.serializeArguments)(options);
        const nonRevokedIntervalOverrideList = this.convertNonRevokedIntervalOverrides(options.nonRevokedIntervalOverrides);
        const ret = (0, ffi_1.allocateInt8Buffer)();
        const errorCode = this.nativeAnoncreds.anoncreds_verify_presentation(presentation, presentationRequest, schemas, schemaIds, credentialDefinitions, credentialDefinitionIds, revocationRegistryDefinitions !== null && revocationRegistryDefinitions !== void 0 ? revocationRegistryDefinitions : {}, revocationRegistryDefinitionIds !== null && revocationRegistryDefinitionIds !== void 0 ? revocationRegistryDefinitionIds : {}, revocationStatusLists !== null && revocationStatusLists !== void 0 ? revocationStatusLists : {}, nonRevokedIntervalOverrideList, ret);
        this.handleError(errorCode);
        return Boolean(handleReturnPointer(ret));
    }
    createRevocationStatusList(options) {
        const { credentialDefinition, revocationRegistryDefinitionId, revocationRegistryDefinition, revocationRegistryDefinitionPrivate, issuerId, issuanceByDefault, timestamp, } = (0, ffi_1.serializeArguments)(options);
        const ret = (0, ffi_1.allocatePointer)();
        const errorCode = this.nativeAnoncreds.anoncreds_create_revocation_status_list(credentialDefinition, revocationRegistryDefinitionId, revocationRegistryDefinition, revocationRegistryDefinitionPrivate, issuerId, issuanceByDefault, timestamp !== null && timestamp !== void 0 ? timestamp : -1, ret);
        this.handleError(errorCode);
        return new anoncreds_shared_1.ObjectHandle(handleReturnPointer(ret));
    }
    updateRevocationStatusListTimestampOnly(options) {
        const { currentRevocationStatusList, timestamp } = (0, ffi_1.serializeArguments)(options);
        const ret = (0, ffi_1.allocatePointer)();
        const errorCode = this.nativeAnoncreds.anoncreds_update_revocation_status_list_timestamp_only(timestamp, currentRevocationStatusList, ret);
        this.handleError(errorCode);
        return new anoncreds_shared_1.ObjectHandle(handleReturnPointer(ret));
    }
    updateRevocationStatusList(options) {
        const { credentialDefinition, revocationRegistryDefinition, revocationRegistryDefinitionPrivate, currentRevocationStatusList, issued, revoked, timestamp, } = (0, ffi_1.serializeArguments)(options);
        const ret = (0, ffi_1.allocatePointer)();
        const errorCode = this.nativeAnoncreds.anoncreds_update_revocation_status_list(credentialDefinition, revocationRegistryDefinition, revocationRegistryDefinitionPrivate, currentRevocationStatusList, issued !== null && issued !== void 0 ? issued : {}, revoked !== null && revoked !== void 0 ? revoked : {}, timestamp !== null && timestamp !== void 0 ? timestamp : -1, ret);
        this.handleError(errorCode);
        return new anoncreds_shared_1.ObjectHandle(handleReturnPointer(ret));
    }
    createRevocationRegistryDefinition(options) {
        const { credentialDefinition, credentialDefinitionId, tag, revocationRegistryType, issuerId, maximumCredentialNumber, tailsDirectoryPath, } = (0, ffi_1.serializeArguments)(options);
        const revocationRegistryDefinitionPtr = (0, ffi_1.allocatePointer)();
        const revocationRegistryDefinitionPrivate = (0, ffi_1.allocatePointer)();
        const errorCode = this.nativeAnoncreds.anoncreds_create_revocation_registry_def(credentialDefinition, credentialDefinitionId, issuerId, tag, revocationRegistryType, maximumCredentialNumber, tailsDirectoryPath, revocationRegistryDefinitionPtr, revocationRegistryDefinitionPrivate);
        this.handleError(errorCode);
        return {
            revocationRegistryDefinition: new anoncreds_shared_1.ObjectHandle(handleReturnPointer(revocationRegistryDefinitionPtr)),
            revocationRegistryDefinitionPrivate: new anoncreds_shared_1.ObjectHandle(handleReturnPointer(revocationRegistryDefinitionPrivate)),
        };
    }
    createOrUpdateRevocationState(options) {
        var _a, _b;
        const { revocationRegistryDefinition, revocationStatusList, revocationRegistryIndex, tailsPath } = (0, ffi_1.serializeArguments)(options);
        const oldRevocationState = (_a = options.oldRevocationState) !== null && _a !== void 0 ? _a : new anoncreds_shared_1.ObjectHandle(0);
        const oldRevocationStatusList = (_b = options.oldRevocationStatusList) !== null && _b !== void 0 ? _b : new anoncreds_shared_1.ObjectHandle(0);
        const ret = (0, ffi_1.allocatePointer)();
        const errorCode = this.nativeAnoncreds.anoncreds_create_or_update_revocation_state(revocationRegistryDefinition, revocationStatusList, revocationRegistryIndex, tailsPath, oldRevocationState.handle, oldRevocationStatusList.handle, ret);
        this.handleError(errorCode);
        return new anoncreds_shared_1.ObjectHandle(handleReturnPointer(ret));
    }
    createW3cCredential(options) {
        const { credentialDefinition, credentialDefinitionPrivate, credentialOffer, credentialRequest, w3cVersion } = (0, ffi_1.serializeArguments)(options);
        const attributeNames = this.convertAttributeNames(options.attributeRawValues);
        const attributeRawValues = this.convertAttributeRawValues(options.attributeRawValues);
        const revocationConfiguration = this.convertRevocationConfiguration(options.revocationConfiguration);
        const credentialPtr = (0, ffi_1.allocatePointer)();
        const errorCode = this.nativeAnoncreds.anoncreds_create_w3c_credential(credentialDefinition, credentialDefinitionPrivate, credentialOffer, credentialRequest, attributeNames, attributeRawValues, revocationConfiguration !== null && revocationConfiguration !== void 0 ? revocationConfiguration : 0, w3cVersion, credentialPtr);
        this.handleError(errorCode);
        return new anoncreds_shared_1.ObjectHandle(handleReturnPointer(credentialPtr));
    }
    processW3cCredential(options) {
        var _a, _b;
        const { credential, credentialRequestMetadata, linkSecret, credentialDefinition } = (0, ffi_1.serializeArguments)(options);
        const ret = (0, ffi_1.allocatePointer)();
        const errorCode = this.nativeAnoncreds.anoncreds_process_w3c_credential(credential, credentialRequestMetadata, linkSecret, credentialDefinition, (_b = (_a = options.revocationRegistryDefinition) === null || _a === void 0 ? void 0 : _a.handle) !== null && _b !== void 0 ? _b : 0, ret);
        this.handleError(errorCode);
        return new anoncreds_shared_1.ObjectHandle(handleReturnPointer(ret));
    }
    createW3cPresentation(options) {
        const { presentationRequest, linkSecret, w3cVersion } = (0, ffi_1.serializeArguments)(options);
        const credentialEntryList = this.convertCredentialList(options.credentials);
        const credentialProveList = this.convertCredentialProves(options.credentialsProve);
        const { schemaIds, schemas } = this.convertSchemas(options.schemas);
        const { credentialDefinitionIds, credentialDefinitions } = this.convertCredDefs(options.credentialDefinitions);
        const ret = (0, ffi_1.allocatePointer)();
        const errorCode = this.nativeAnoncreds.anoncreds_create_w3c_presentation(presentationRequest, credentialEntryList, credentialProveList, linkSecret, schemas, schemaIds, credentialDefinitions, credentialDefinitionIds, w3cVersion, ret);
        this.handleError(errorCode);
        return new anoncreds_shared_1.ObjectHandle(handleReturnPointer(ret));
    }
    verifyW3cPresentation(options) {
        const { presentation, presentationRequest, schemas, credentialDefinitions, revocationRegistryDefinitions, revocationStatusLists, revocationRegistryDefinitionIds, schemaIds, credentialDefinitionIds, } = (0, ffi_1.serializeArguments)(options);
        const nonRevokedIntervalOverrideList = this.convertNonRevokedIntervalOverrides(options.nonRevokedIntervalOverrides);
        const ret = (0, ffi_1.allocateInt8Buffer)();
        const errorCode = this.nativeAnoncreds.anoncreds_verify_w3c_presentation(presentation, presentationRequest, schemas, schemaIds, credentialDefinitions, credentialDefinitionIds, revocationRegistryDefinitions !== null && revocationRegistryDefinitions !== void 0 ? revocationRegistryDefinitions : {}, revocationRegistryDefinitionIds !== null && revocationRegistryDefinitionIds !== void 0 ? revocationRegistryDefinitionIds : {}, revocationStatusLists !== null && revocationStatusLists !== void 0 ? revocationStatusLists : {}, nonRevokedIntervalOverrideList, ret);
        this.handleError(errorCode);
        return Boolean(handleReturnPointer(ret));
    }
    credentialToW3c(options) {
        const { objectHandle, issuerId, w3cVersion } = (0, ffi_1.serializeArguments)(options);
        const ret = (0, ffi_1.allocatePointer)();
        const errorCode = this.nativeAnoncreds.anoncreds_credential_to_w3c(objectHandle, issuerId, w3cVersion, ret);
        this.handleError(errorCode);
        return new anoncreds_shared_1.ObjectHandle(handleReturnPointer(ret));
    }
    credentialFromW3c(options) {
        const { objectHandle } = (0, ffi_1.serializeArguments)(options);
        const ret = (0, ffi_1.allocatePointer)();
        const errorCode = this.nativeAnoncreds.anoncreds_credential_from_w3c(objectHandle, ret);
        this.handleError(errorCode);
        return new anoncreds_shared_1.ObjectHandle(handleReturnPointer(ret));
    }
    w3cCredentialGetIntegrityProofDetails(options) {
        const { objectHandle } = (0, ffi_1.serializeArguments)(options);
        const ret = (0, ffi_1.allocatePointer)();
        const errorCode = this.nativeAnoncreds.anoncreds_w3c_credential_get_integrity_proof_details(objectHandle, ret);
        this.handleError(errorCode);
        return new anoncreds_shared_1.ObjectHandle(handleReturnPointer(ret));
    }
    w3cCredentialProofGetAttribute(options) {
        var _a;
        const { objectHandle, name } = (0, ffi_1.serializeArguments)(options);
        const ret = (0, ffi_1.allocateStringBuffer)();
        const errorCode = this.nativeAnoncreds.anoncreds_w3c_credential_proof_get_attribute(objectHandle, name, ret);
        this.handleError(errorCode);
        const value = Array.isArray(ret) ? ret[0] : ret;
        return (_a = value) !== null && _a !== void 0 ? _a : null;
    }
    w3cCredentialFromJson(options) {
        return this.objectFromJson(this.nativeAnoncreds.anoncreds_w3c_credential_from_json, options);
    }
    w3cPresentationFromJson(options) {
        return this.objectFromJson(this.nativeAnoncreds.anoncreds_w3c_presentation_from_json, options);
    }
    version() {
        return this.nativeAnoncreds.anoncreds_version();
    }
    setDefaultLogger() {
        const errorCode = this.nativeAnoncreds.anoncreds_set_default_logger();
        this.handleError(errorCode);
    }
    objectFromJson(cb, options) {
        const ret = (0, ffi_1.allocatePointer)();
        const byteBuffer = new node_util_1.TextEncoder().encode(options.json);
        const errorCode = cb({ data: byteBuffer, len: byteBuffer.length }, ret);
        this.handleError(errorCode);
        return new anoncreds_shared_1.ObjectHandle(handleReturnPointer(ret));
    }
    presentationRequestFromJson(options) {
        return this.objectFromJson(this.nativeAnoncreds.anoncreds_presentation_request_from_json, options);
    }
    credentialRequestFromJson(options) {
        return this.objectFromJson(this.nativeAnoncreds.anoncreds_credential_request_from_json, options);
    }
    credentialRequestMetadataFromJson(options) {
        return this.objectFromJson(this.nativeAnoncreds.anoncreds_credential_request_metadata_from_json, options);
    }
    revocationRegistryDefinitionFromJson(options) {
        return this.objectFromJson(this.nativeAnoncreds.anoncreds_revocation_registry_definition_from_json, options);
    }
    revocationRegistryFromJson(options) {
        return this.objectFromJson(this.nativeAnoncreds.anoncreds_revocation_registry_from_json, options);
    }
    revocationStatusListFromJson(options) {
        return this.objectFromJson(this.nativeAnoncreds.anoncreds_revocation_status_list_from_json, options);
    }
    revocationStateFromJson(options) {
        return this.objectFromJson(this.nativeAnoncreds.anoncreds_revocation_state_from_json, options);
    }
    presentationFromJson(options) {
        return this.objectFromJson(this.nativeAnoncreds.anoncreds_presentation_from_json, options);
    }
    credentialOfferFromJson(options) {
        return this.objectFromJson(this.nativeAnoncreds.anoncreds_credential_offer_from_json, options);
    }
    schemaFromJson(options) {
        return this.objectFromJson(this.nativeAnoncreds.anoncreds_schema_from_json, options);
    }
    credentialFromJson(options) {
        return this.objectFromJson(this.nativeAnoncreds.anoncreds_credential_from_json, options);
    }
    revocationRegistryDefinitionPrivateFromJson(options) {
        return this.objectFromJson(this.nativeAnoncreds.anoncreds_revocation_registry_definition_private_from_json, options);
    }
    credentialDefinitionFromJson(options) {
        return this.objectFromJson(this.nativeAnoncreds.anoncreds_credential_definition_from_json, options);
    }
    credentialDefinitionPrivateFromJson(options) {
        return this.objectFromJson(this.nativeAnoncreds.anoncreds_credential_definition_private_from_json, options);
    }
    keyCorrectnessProofFromJson(options) {
        return this.objectFromJson(this.nativeAnoncreds.anoncreds_key_correctness_proof_from_json, options);
    }
    getJson(options) {
        const { objectHandle } = (0, ffi_1.serializeArguments)(options);
        const ret = (0, ffi_1.allocateByteBuffer)();
        const errorCode = this.nativeAnoncreds.anoncreds_object_get_json(objectHandle, ret);
        this.handleError(errorCode);
        const returnValue = handleReturnPointer(ret);
        const jsonAsArray = (0, ffi_1.byteBufferToUint8Array)(returnValue);
        const output = new node_util_1.TextDecoder().decode(jsonAsArray);
        this.nativeAnoncreds.anoncreds_buffer_free(returnValue);
        return output;
    }
    getTypeName(options) {
        const { objectHandle } = (0, ffi_1.serializeArguments)(options);
        const ret = (0, ffi_1.allocateStringBuffer)();
        const errorCode = this.nativeAnoncreds.anoncreds_object_get_type_name(objectHandle, ret);
        this.handleError(errorCode);
        return handleReturnPointer(ret);
    }
    objectFree(options) {
        this.nativeAnoncreds.anoncreds_object_free(options.objectHandle.handle);
    }
    convertCredentialList(credentials) {
        const credentialEntries = credentials.map((value) => {
            var _a, _b, _c;
            return ({
                credential: value.credential.handle,
                timestamp: (_a = value.timestamp) !== null && _a !== void 0 ? _a : -1,
                rev_state: (_c = (_b = value.revocationState) === null || _b === void 0 ? void 0 : _b.handle) !== null && _c !== void 0 ? _c : 0,
            });
        });
        return {
            count: credentialEntries.length,
            data: credentialEntries,
        };
    }
    convertCredentialProves(credentialsProve) {
        const credentialProves = credentialsProve.map((value) => {
            const { entryIndex: entry_idx, isPredicate: is_predicate, reveal, referent } = (0, ffi_1.serializeArguments)(value);
            return { entry_idx, referent, is_predicate, reveal };
        });
        return {
            count: credentialProves.length,
            data: credentialProves,
        };
    }
    convertSchemas(schemas) {
        const schemaKeys = Object.keys(schemas);
        const schemaIds = {
            count: schemaKeys.length,
            data: schemaKeys,
        };
        const schemaValues = Object.values(schemas);
        return {
            schemaIds,
            schemas: {
                count: schemaValues.length,
                data: schemaValues.map((o) => o.handle),
            },
        };
    }
    convertCredDefs(credentialDefinitions) {
        const credentialDefinitionKeys = Object.keys(credentialDefinitions);
        const credentialDefinitionIds = {
            count: credentialDefinitionKeys.length,
            data: credentialDefinitionKeys,
        };
        const credentialDefinitionValues = Object.values(credentialDefinitions);
        const credentialDefinitionsList = {
            count: credentialDefinitionValues.length,
            data: credentialDefinitionValues.map((o) => o.handle),
        };
        return {
            credentialDefinitionIds,
            credentialDefinitions: credentialDefinitionsList,
        };
    }
    convertNonRevokedIntervalOverrides(nonRevokedIntervalOverrides) {
        var _a;
        const nativeNonRevokedIntervalOverrides = nonRevokedIntervalOverrides === null || nonRevokedIntervalOverrides === void 0 ? void 0 : nonRevokedIntervalOverrides.map((value) => {
            const { requestedFromTimestamp, revocationRegistryDefinitionId, overrideRevocationStatusListTimestamp } = (0, ffi_1.serializeArguments)(value);
            return {
                rev_reg_def_id: revocationRegistryDefinitionId,
                requested_from_ts: requestedFromTimestamp,
                override_rev_status_list_ts: overrideRevocationStatusListTimestamp,
            };
        });
        return {
            count: (_a = nonRevokedIntervalOverrides === null || nonRevokedIntervalOverrides === void 0 ? void 0 : nonRevokedIntervalOverrides.length) !== null && _a !== void 0 ? _a : 0,
            data: nativeNonRevokedIntervalOverrides,
        };
    }
    convertAttributeNames(attributeRawValues) {
        return {
            count: Object.keys(attributeRawValues).length,
            data: Object.keys(attributeRawValues),
        };
    }
    convertAttributeRawValues(attributeRawValues) {
        return {
            count: Object.keys(attributeRawValues).length,
            data: Object.values(attributeRawValues),
        };
    }
    convertAttributeEncodedValues(attributeEncodedValues) {
        return attributeEncodedValues
            ? {
                count: Object.keys(attributeEncodedValues).length,
                data: Object.values(attributeEncodedValues),
            }
            : undefined;
    }
    convertRevocationConfiguration(revocationConfiguration) {
        if (revocationConfiguration) {
            const { revocationRegistryDefinition, revocationRegistryDefinitionPrivate, revocationStatusList, registryIndex } = (0, ffi_1.serializeArguments)(revocationConfiguration);
            return {
                reg_def: revocationRegistryDefinition,
                reg_def_private: revocationRegistryDefinitionPrivate,
                status_list: revocationStatusList,
                reg_idx: registryIndex,
            };
        }
    }
}
exports.NodeJSAnoncreds = NodeJSAnoncreds;
