// @sphereon/ssi-types/dist/types/cose.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ICoseCurve = exports.ICoseKeyOperation = exports.ICoseSignatureAlgorithm = exports.ICoseKeyType = void 0;
var ICoseKeyType;
(function (ICoseKeyType) {
    ICoseKeyType[ICoseKeyType["OKP"] = 1] = "OKP";
    ICoseKeyType[ICoseKeyType["EC2"] = 2] = "EC2";
    ICoseKeyType[ICoseKeyType["RSA"] = 3] = "RSA";
    ICoseKeyType[ICoseKeyType["Symmetric"] = 4] = "Symmetric";
    ICoseKeyType[ICoseKeyType["Reserved"] = 0] = "Reserved";
})(ICoseKeyType || (exports.ICoseKeyType = ICoseKeyType = {}));
var ICoseSignatureAlgorithm;
(function (ICoseSignatureAlgorithm) {
    ICoseSignatureAlgorithm[ICoseSignatureAlgorithm["ES256"] = -7] = "ES256";
    ICoseSignatureAlgorithm[ICoseSignatureAlgorithm["ES256K"] = -47] = "ES256K";
    ICoseSignatureAlgorithm[ICoseSignatureAlgorithm["ES384"] = -35] = "ES384";
    ICoseSignatureAlgorithm[ICoseSignatureAlgorithm["ES512"] = -36] = "ES512";
    ICoseSignatureAlgorithm[ICoseSignatureAlgorithm["EdDSA"] = -8] = "EdDSA";
    ICoseSignatureAlgorithm[ICoseSignatureAlgorithm["HS256_64"] = 4] = "HS256_64";
    ICoseSignatureAlgorithm[ICoseSignatureAlgorithm["HS256"] = 5] = "HS256";
    ICoseSignatureAlgorithm[ICoseSignatureAlgorithm["HS384"] = 6] = "HS384";
    ICoseSignatureAlgorithm[ICoseSignatureAlgorithm["HS512"] = 7] = "HS512";
    ICoseSignatureAlgorithm[ICoseSignatureAlgorithm["PS256"] = -37] = "PS256";
    ICoseSignatureAlgorithm[ICoseSignatureAlgorithm["PS384"] = -38] = "PS384";
    ICoseSignatureAlgorithm[ICoseSignatureAlgorithm["PS512"] = -39] = "PS512";
})(ICoseSignatureAlgorithm || (exports.ICoseSignatureAlgorithm = ICoseSignatureAlgorithm = {}));
var ICoseKeyOperation;
(function (ICoseKeyOperation) {
    ICoseKeyOperation[ICoseKeyOperation["SIGN"] = 1] = "SIGN";
    ICoseKeyOperation[ICoseKeyOperation["VERIFY"] = 2] = "VERIFY";
    ICoseKeyOperation[ICoseKeyOperation["ENCRYPT"] = 3] = "ENCRYPT";
    ICoseKeyOperation[ICoseKeyOperation["DECRYPT"] = 4] = "DECRYPT";
    ICoseKeyOperation[ICoseKeyOperation["WRAP_KEY"] = 5] = "WRAP_KEY";
    ICoseKeyOperation[ICoseKeyOperation["UNWRAP_KEY"] = 6] = "UNWRAP_KEY";
    ICoseKeyOperation[ICoseKeyOperation["DERIVE_KEY"] = 7] = "DERIVE_KEY";
    ICoseKeyOperation[ICoseKeyOperation["DERIVE_BITS"] = 8] = "DERIVE_BITS";
    ICoseKeyOperation[ICoseKeyOperation["MAC_CREATE"] = 9] = "MAC_CREATE";
    ICoseKeyOperation[ICoseKeyOperation["MAC_VERIFY"] = 10] = "MAC_VERIFY";
})(ICoseKeyOperation || (exports.ICoseKeyOperation = ICoseKeyOperation = {}));
var ICoseCurve;
(function (ICoseCurve) {
    ICoseCurve[ICoseCurve["P_256"] = 1] = "P_256";
    ICoseCurve[ICoseCurve["P_384"] = 2] = "P_384";
    ICoseCurve[ICoseCurve["P_521"] = 3] = "P_521";
    ICoseCurve[ICoseCurve["X25519"] = 4] = "X25519";
    ICoseCurve[ICoseCurve["X448"] = 5] = "X448";
    ICoseCurve[ICoseCurve["Ed25519"] = 6] = "Ed25519";
    ICoseCurve[ICoseCurve["Ed448"] = 7] = "Ed448";
    ICoseCurve[ICoseCurve["secp256k1"] = -1] = "secp256k1";
})(ICoseCurve || (exports.ICoseCurve = ICoseCurve = {}));
