// @sphereon/ssi-types/dist/types/jose.js
"use strict";
/**
 * Conversion functions to Cose available for TS in our @sphereon/ssi-sdk-ext.key-utils package
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.JoseCurve = exports.JoseKeyOperation = exports.JoseSignatureAlgorithm = exports.JwkKeyType = void 0;
var JwkKeyType;
(function (JwkKeyType) {
    JwkKeyType["EC"] = "EC";
    JwkKeyType["RSA"] = "RSA";
    JwkKeyType["oct"] = "oct";
    JwkKeyType["OKP"] = "OKP";
})(JwkKeyType || (exports.JwkKeyType = JwkKeyType = {}));
var JoseSignatureAlgorithm;
(function (JoseSignatureAlgorithm) {
    JoseSignatureAlgorithm["RS256"] = "RS256";
    JoseSignatureAlgorithm["RS384"] = "RS384";
    JoseSignatureAlgorithm["RS512"] = "RS512";
    JoseSignatureAlgorithm["ES256"] = "ES256";
    JoseSignatureAlgorithm["ES256K"] = "ES256K";
    JoseSignatureAlgorithm["ES384"] = "ES384";
    JoseSignatureAlgorithm["ES512"] = "ES512";
    JoseSignatureAlgorithm["EdDSA"] = "EdDSA";
    JoseSignatureAlgorithm["HS256"] = "HS256";
    JoseSignatureAlgorithm["HS384"] = "HS384";
    JoseSignatureAlgorithm["HS512"] = "HS512";
    JoseSignatureAlgorithm["PS256"] = "PS256";
    JoseSignatureAlgorithm["PS384"] = "PS384";
    JoseSignatureAlgorithm["PS512"] = "PS512";
    JoseSignatureAlgorithm["none"] = "none";
})(JoseSignatureAlgorithm || (exports.JoseSignatureAlgorithm = JoseSignatureAlgorithm = {}));
var JoseKeyOperation;
(function (JoseKeyOperation) {
    JoseKeyOperation["SIGN"] = "sign";
    JoseKeyOperation["VERIFY"] = "verify";
    JoseKeyOperation["ENCRYPT"] = "encrypt";
    JoseKeyOperation["DECRYPT"] = "decrypt";
    JoseKeyOperation["WRAP_KEY"] = "wrapKey";
    JoseKeyOperation["UNWRAP_KEY"] = "unwrapKey";
    JoseKeyOperation["DERIVE_KEY"] = "deriveKey";
    JoseKeyOperation["DERIVE_BITS"] = "deriveBits";
})(JoseKeyOperation || (exports.JoseKeyOperation = JoseKeyOperation = {}));
var JoseCurve;
(function (JoseCurve) {
    JoseCurve["P_256"] = "P-256";
    JoseCurve["P_384"] = "P-384";
    JoseCurve["P_521"] = "P-521";
    JoseCurve["X25519"] = "X25519";
    JoseCurve["X448"] = "X448";
    JoseCurve["EdDSA"] = "EdDSA";
    JoseCurve["Ed25519"] = "Ed25519";
    JoseCurve["Ed448"] = "Ed448";
    JoseCurve["secp256k1"] = "secp256k1";
})(JoseCurve || (exports.JoseCurve = JoseCurve = {}));
