// @hyperledger/aries-askar-shared/build/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AeadParams = exports.EncryptedBuffer = void 0;
class EncryptedBuffer {
    constructor({ noncePos, tagPos, buffer }) {
        this.buffer = buffer;
        this.tagPos = tagPos;
        this.noncePos = noncePos;
    }
    get ciphertextWithTag() {
        const p = this.noncePos;
        return this.buffer.slice(0, p);
    }
    get ciphertext() {
        const p = this.tagPos;
        return this.buffer.slice(0, p);
    }
    get nonce() {
        const p = this.noncePos;
        return this.buffer.slice(p);
    }
    get tag() {
        const p1 = this.tagPos;
        const p2 = this.noncePos;
        return this.buffer.slice(p1, p2);
    }
    get parts() {
        return {
            ciphertext: this.ciphertext,
            tag: this.tag,
            nonce: this.nonce,
        };
    }
}
exports.EncryptedBuffer = EncryptedBuffer;
class AeadParams {
    constructor({ nonceLength, tagLength }) {
        this.nonceLength = nonceLength;
        this.tagLength = tagLength;
    }
}
exports.AeadParams = AeadParams;
