// @hyperledger/anoncreds-shared/build/register.js
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _a, _NativeAnoncreds_nativeAnoncreds;
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerAnoncreds = exports.NativeAnoncreds = exports.anoncreds = void 0;
const error_1 = require("./error");
class NativeAnoncreds {
    static get instance() {
        if (!__classPrivateFieldGet(_a, _a, "f", _NativeAnoncreds_nativeAnoncreds))
            throw error_1.AnoncredsError.customError({
                message: "Native anoncreds has not been registered yet. Make sure to import '@hyperledger/anoncreds-nodejs' or '@hyperledger/anoncreds-react-native', or call 'NativeAnoncreds.register' with a custom implementation.",
            });
        return __classPrivateFieldGet(_a, _a, "f", _NativeAnoncreds_nativeAnoncreds);
    }
    static register(nativeAnoncreds) {
        exports.anoncreds = nativeAnoncreds;
        __classPrivateFieldSet(_a, _a, nativeAnoncreds, "f", _NativeAnoncreds_nativeAnoncreds);
    }
}
exports.NativeAnoncreds = NativeAnoncreds;
_a = NativeAnoncreds;
_NativeAnoncreds_nativeAnoncreds = { value: void 0 };
/**
 * @deprecated use `NativeAnoncreds.register` instead
 */
const registerAnoncreds = ({ lib }) => {
    exports.anoncreds = lib;
    NativeAnoncreds.register(lib);
};
exports.registerAnoncreds = registerAnoncreds;
