// @animo-id/pex/dist/browser/lib/evaluation/handlers/markForSubmissionEvaluationHandler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MarkForSubmissionEvaluationHandler = exports.eligibleInputDescriptorsForWrappedVc = void 0;
const jsonpath_1 = require("@astronautlabs/jsonpath");
const ConstraintUtils_1 = require("../../ConstraintUtils");
const Messages_1 = __importDefault(require("../../types/Messages"));
const abstractEvaluationHandler_1 = require("./abstractEvaluationHandler");
function eligibleInputDescriptorsForWrappedVc(inputDescriptors, vcIndex, results) {
    return inputDescriptors
        .map((inputDescriptor, inputDescriptorIndex) => {
        const matchingResults = results.filter(({ verifiable_credential_path, input_descriptor_path }) => verifiable_credential_path === `$[${vcIndex}]` && input_descriptor_path === `$.input_descriptors[${inputDescriptorIndex}]`);
        const hasError = matchingResults.some((result) => result.status === ConstraintUtils_1.Status.ERROR);
        const hasInfo = matchingResults.some((result) => result.status === ConstraintUtils_1.Status.INFO);
        if (hasInfo && !hasError) {
            return {
                inputDescriptor,
                inputDescriptorIndex,
            };
        }
        return undefined;
    })
        .filter((value) => value !== undefined);
}
exports.eligibleInputDescriptorsForWrappedVc = eligibleInputDescriptorsForWrappedVc;
class MarkForSubmissionEvaluationHandler extends abstractEvaluationHandler_1.AbstractEvaluationHandler {
    constructor(client) {
        super(client);
    }
    getName() {
        return 'MarkForSubmissionEvaluation';
    }
    handle(pd, wrappedVcs) {
        const results = [...this.getResults()];
        const errors = results.filter((result) => result.status === ConstraintUtils_1.Status.ERROR);
        const infos = this.retrieveNoErrorStatus(results, errors);
        this.client.wrappedVcs = wrappedVcs;
        this.produceErrorResults(errors);
        this.produceSuccessResults(infos, pd);
    }
    retrieveNoErrorStatus(results, errors) {
        const info = results.filter((e) => e.status !== ConstraintUtils_1.Status.ERROR);
        return info.filter((a) => !errors.find((b) => a.input_descriptor_path === b.input_descriptor_path && a.verifiable_credential_path === b.verifiable_credential_path));
    }
    produceSuccessResults(infos, pd) {
        this.removeDuplicate(infos).forEach((info) => {
            const parsedPath = jsonpath_1.JSONPath.nodes(pd, info.input_descriptor_path);
            const group = parsedPath[0].value.group;
            this.getResults().push({
                input_descriptor_path: info.input_descriptor_path,
                verifiable_credential_path: info.verifiable_credential_path,
                evaluator: this.getName(),
                status: ConstraintUtils_1.Status.INFO,
                payload: { group },
                message: Messages_1.default.INPUT_CANDIDATE_IS_ELIGIBLE_FOR_PRESENTATION_SUBMISSION,
            });
        });
    }
    produceErrorResults(errors) {
        this.removeDuplicate(errors).forEach((error) => {
            const payload = Object.assign({}, error.payload);
            payload.evaluator = error.evaluator;
            this.getResults().push(Object.assign(Object.assign({}, error), { evaluator: this.getName(), message: Messages_1.default.INPUT_CANDIDATE_IS_NOT_ELIGIBLE_FOR_PRESENTATION_SUBMISSION, payload: payload }));
        });
    }
}
exports.MarkForSubmissionEvaluationHandler = MarkForSubmissionEvaluationHandler;
