// @credo-ts/core/build/modules/dids/domain/DidDocumentBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DidDocumentBuilder = void 0;
const utils_1 = require("../../../utils");
const DidDocument_1 = require("./DidDocument");
const verificationMethod_1 = require("./verificationMethod");
class DidDocumentBuilder {
    constructor(id) {
        this.didDocument = new DidDocument_1.DidDocument({
            id,
        });
    }
    addContext(context) {
        const currentContexts = (0, utils_1.asArray)(this.didDocument.context);
        if (currentContexts.includes(context))
            return this;
        this.didDocument.context = [...currentContexts, context];
        return this;
    }
    addController(controller) {
        const currentControllers = (0, utils_1.asArray)(this.didDocument.controller);
        if (currentControllers.includes(controller))
            return this;
        this.didDocument.controller = [...currentControllers, controller];
        return this;
    }
    addService(service) {
        if (!this.didDocument.service) {
            this.didDocument.service = [];
        }
        this.didDocument.service.push(service);
        return this;
    }
    addVerificationMethod(verificationMethod) {
        if (!this.didDocument.verificationMethod) {
            this.didDocument.verificationMethod = [];
        }
        this.didDocument.verificationMethod.push(verificationMethod instanceof verificationMethod_1.VerificationMethod ? verificationMethod : new verificationMethod_1.VerificationMethod(verificationMethod));
        return this;
    }
    addAuthentication(authentication) {
        if (!this.didDocument.authentication) {
            this.didDocument.authentication = [];
        }
        const verificationMethod = authentication instanceof verificationMethod_1.VerificationMethod || typeof authentication === 'string'
            ? authentication
            : new verificationMethod_1.VerificationMethod(authentication);
        this.didDocument.authentication.push(verificationMethod);
        return this;
    }
    addAssertionMethod(assertionMethod) {
        if (!this.didDocument.assertionMethod) {
            this.didDocument.assertionMethod = [];
        }
        const verificationMethod = assertionMethod instanceof verificationMethod_1.VerificationMethod || typeof assertionMethod === 'string'
            ? assertionMethod
            : new verificationMethod_1.VerificationMethod(assertionMethod);
        this.didDocument.assertionMethod.push(verificationMethod);
        return this;
    }
    addCapabilityDelegation(capabilityDelegation) {
        if (!this.didDocument.capabilityDelegation) {
            this.didDocument.capabilityDelegation = [];
        }
        const verificationMethod = capabilityDelegation instanceof verificationMethod_1.VerificationMethod || typeof capabilityDelegation === 'string'
            ? capabilityDelegation
            : new verificationMethod_1.VerificationMethod(capabilityDelegation);
        this.didDocument.capabilityDelegation.push(verificationMethod);
        return this;
    }
    addCapabilityInvocation(capabilityInvocation) {
        if (!this.didDocument.capabilityInvocation) {
            this.didDocument.capabilityInvocation = [];
        }
        const verificationMethod = capabilityInvocation instanceof verificationMethod_1.VerificationMethod || typeof capabilityInvocation === 'string'
            ? capabilityInvocation
            : new verificationMethod_1.VerificationMethod(capabilityInvocation);
        this.didDocument.capabilityInvocation.push(verificationMethod);
        return this;
    }
    addKeyAgreement(keyAgreement) {
        if (!this.didDocument.keyAgreement) {
            this.didDocument.keyAgreement = [];
        }
        const verificationMethod = keyAgreement instanceof verificationMethod_1.VerificationMethod || typeof keyAgreement === 'string'
            ? keyAgreement
            : new verificationMethod_1.VerificationMethod(keyAgreement);
        this.didDocument.keyAgreement.push(verificationMethod);
        return this;
    }
    build() {
        return this.didDocument;
    }
}
exports.DidDocumentBuilder = DidDocumentBuilder;
