// @credo-ts/core/build/modules/dids/repository/DidRecord.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DidRecord = void 0;
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const BaseRecord_1 = require("../../../storage/BaseRecord");
const uuid_1 = require("../../../utils/uuid");
const domain_1 = require("../domain");
const DidDocumentRole_1 = require("../domain/DidDocumentRole");
const parse_1 = require("../domain/parse");
const didRecordMetadataTypes_1 = require("./didRecordMetadataTypes");
class DidRecord extends BaseRecord_1.BaseRecord {
    constructor(props) {
        super();
        this.type = DidRecord.type;
        if (props) {
            this.id = props.id ?? (0, uuid_1.uuid)();
            this.did = props.did;
            this.role = props.role;
            this.didDocument = props.didDocument;
            this.createdAt = props.createdAt ?? new Date();
            this._tags = props.tags ?? {};
        }
    }
    getTags() {
        const did = (0, parse_1.parseDid)(this.did);
        const legacyDid = this.metadata.get(didRecordMetadataTypes_1.DidRecordMetadataKeys.LegacyDid);
        return {
            ...this._tags,
            role: this.role,
            method: did.method,
            legacyUnqualifiedDid: legacyDid?.unqualifiedDid,
            did: this.did,
            methodSpecificIdentifier: did.id,
            // Calculate if we have a did document, otherwise use the already present recipient keys
            recipientKeyFingerprints: this.didDocument
                ? this.didDocument.recipientKeys.map((recipientKey) => recipientKey.fingerprint)
                : this._tags.recipientKeyFingerprints,
        };
    }
}
exports.DidRecord = DidRecord;
DidRecord.type = 'DidRecord';
__decorate([
    (0, class_transformer_1.Type)(() => domain_1.DidDocument),
    (0, class_validator_1.ValidateNested)(),
    __metadata("design:type", domain_1.DidDocument)
], DidRecord.prototype, "didDocument", void 0);
__decorate([
    (0, class_validator_1.IsEnum)(DidDocumentRole_1.DidDocumentRole),
    __metadata("design:type", String)
], DidRecord.prototype, "role", void 0);
