// @credo-ts/askar/build/AskarModuleConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AskarModuleConfig = exports.AskarMultiWalletDatabaseScheme = void 0;
const importAskar_1 = require("./utils/importAskar");
var AskarMultiWalletDatabaseScheme;
(function (AskarMultiWalletDatabaseScheme) {
    /**
     * Each wallet get its own database and uses a separate store.
     */
    AskarMultiWalletDatabaseScheme["DatabasePerWallet"] = "DatabasePerWallet";
    /**
     * All wallets are stored in a single database, but each wallet uses a separate profile.
     */
    AskarMultiWalletDatabaseScheme["ProfilePerWallet"] = "ProfilePerWallet";
})(AskarMultiWalletDatabaseScheme || (exports.AskarMultiWalletDatabaseScheme = AskarMultiWalletDatabaseScheme = {}));
/**
 * @public
 */
class AskarModuleConfig {
    constructor(options) {
        this.options = options;
        this.askarLibrary = (0, importAskar_1.importAskar)(this.options.ariesAskar);
    }
    /** See {@link AskarModuleConfigOptions.ariesAskar} */
    get ariesAskar() {
        return this.options.ariesAskar;
    }
    /** See {@link AskarModuleConfigOptions.multiWalletDatabaseScheme} */
    get multiWalletDatabaseScheme() {
        return this.options.multiWalletDatabaseScheme ?? AskarMultiWalletDatabaseScheme.DatabasePerWallet;
    }
}
exports.AskarModuleConfig = AskarModuleConfig;
