// @credo-ts/askar/build/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isOwfAskarLibrary = exports.isHyperledgerAskarLibrary = exports.AskarMultiWalletDatabaseScheme = exports.AskarModule = exports.AskarStorageService = exports.AskarProfileWallet = exports.AskarWallet = void 0;
// Wallet
var wallet_1 = require("./wallet");
Object.defineProperty(exports, "AskarWallet", { enumerable: true, get: function () { return wallet_1.AskarWallet; } });
Object.defineProperty(exports, "AskarProfileWallet", { enumerable: true, get: function () { return wallet_1.AskarProfileWallet; } });
// Storage
var storage_1 = require("./storage");
Object.defineProperty(exports, "AskarStorageService", { enumerable: true, get: function () { return storage_1.AskarStorageService; } });
// Module
var AskarModule_1 = require("./AskarModule");
Object.defineProperty(exports, "AskarModule", { enumerable: true, get: function () { return AskarModule_1.AskarModule; } });
var AskarModuleConfig_1 = require("./AskarModuleConfig");
Object.defineProperty(exports, "AskarMultiWalletDatabaseScheme", { enumerable: true, get: function () { return AskarModuleConfig_1.AskarMultiWalletDatabaseScheme; } });
var importAskar_1 = require("./utils/importAskar");
Object.defineProperty(exports, "isHyperledgerAskarLibrary", { enumerable: true, get: function () { return importAskar_1.isHyperledgerAskarLibrary; } });
Object.defineProperty(exports, "isOwfAskarLibrary", { enumerable: true, get: function () { return importAskar_1.isOwfAskarLibrary; } });
