// @hyperledger/anoncreds-nodejs/build/ffi/alloc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.allocateStringListHandle = exports.allocateLocalKeyHandle = exports.allocateAeadParams = exports.allocateEncryptedBuffer = exports.allocateSecretBuffer = exports.allocateByteBuffer = exports.allocateInt8Buffer = exports.allocateInt32Buffer = exports.allocateStringBuffer = exports.allocatePointer = void 0;
const allocatePointer = () => {
    return [undefined];
};
exports.allocatePointer = allocatePointer;
const allocateStringBuffer = () => {
    return [undefined];
};
exports.allocateStringBuffer = allocateStringBuffer;
const allocateInt32Buffer = () => {
    return [0];
};
exports.allocateInt32Buffer = allocateInt32Buffer;
const allocateInt8Buffer = () => {
    return [0];
};
exports.allocateInt8Buffer = allocateInt8Buffer;
const allocateByteBuffer = () => {
    return {};
};
exports.allocateByteBuffer = allocateByteBuffer;
const allocateSecretBuffer = () => {
    return {};
};
exports.allocateSecretBuffer = allocateSecretBuffer;
const allocateEncryptedBuffer = () => {
    return {};
};
exports.allocateEncryptedBuffer = allocateEncryptedBuffer;
const allocateAeadParams = () => {
    return {};
};
exports.allocateAeadParams = allocateAeadParams;
exports.allocateLocalKeyHandle = exports.allocatePointer;
exports.allocateStringListHandle = exports.allocatePointer;
