// @hyperledger/anoncreds-nodejs/build/ffi/conversion.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.i32ListToI32ListStruct = exports.objectHandleListToObjectHandleListStruct = exports.byteBufferToUint8Array = exports.stringListToStringListStruct = void 0;
const koffi_1 = __importDefault(require("koffi"));
const primitives_1 = require("./primitives");
const stringListToStringListStruct = (stringList) => {
    return {
        count: stringList.length,
        data: stringList,
    };
};
exports.stringListToStringListStruct = stringListToStringListStruct;
const byteBufferToUint8Array = (byteBuffer) => {
    const { data, len } = byteBuffer;
    if (data instanceof Uint8Array) {
        return data.slice(0, len);
    }
    const decoded = koffi_1.default.decode(data, primitives_1.FFI_UINT8, len);
    return decoded;
};
exports.byteBufferToUint8Array = byteBufferToUint8Array;
const objectHandleListToObjectHandleListStruct = (objectHandleList) => {
    return {
        count: objectHandleList.length,
        data: objectHandleList.map((h) => h.handle),
    };
};
exports.objectHandleListToObjectHandleListStruct = objectHandleListToObjectHandleListStruct;
const i32ListToI32ListStruct = (i32List) => {
    return {
        count: i32List.length,
        data: i32List,
    };
};
exports.i32ListToI32ListStruct = i32ListToI32ListStruct;
