// @hyperledger/anoncreds-nodejs/build/ffi/structures.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NonrevokedIntervalOverrideStructList = exports.NonrevokedIntervalOverrideStruct = exports.ObjectHandleStructList = exports.CredentialProveStructList = exports.CredentialProveStruct = exports.CredentialEntryStructList = exports.CredentialEntryStruct = exports.CredRevInfoStruct = exports.I32ListStruct = exports.StringListStruct = exports.ByteBufferStruct = void 0;
const koffi_1 = __importDefault(require("koffi"));
const primitives_1 = require("./primitives");
exports.ByteBufferStruct = koffi_1.default.struct('ByteBufferStruct', {
    len: primitives_1.FFI_INT64,
    data: koffi_1.default.pointer(primitives_1.FFI_UINT8),
});
exports.StringListStruct = koffi_1.default.struct('StringListStruct', {
    count: primitives_1.FFI_USIZE,
    data: koffi_1.default.pointer(primitives_1.FFI_STRING),
});
exports.I32ListStruct = koffi_1.default.struct('I32ListStruct', {
    count: primitives_1.FFI_USIZE,
    data: koffi_1.default.pointer(primitives_1.FFI_INT32),
});
exports.CredRevInfoStruct = koffi_1.default.struct('CredRevInfoStruct', {
    reg_def: primitives_1.FFI_OBJECT_HANDLE,
    reg_def_private: primitives_1.FFI_OBJECT_HANDLE,
    status_list: primitives_1.FFI_OBJECT_HANDLE,
    reg_idx: primitives_1.FFI_INT64,
});
exports.CredentialEntryStruct = koffi_1.default.struct('CredentialEntryStruct', {
    credential: primitives_1.FFI_OBJECT_HANDLE,
    timestamp: primitives_1.FFI_INT32,
    rev_state: primitives_1.FFI_OBJECT_HANDLE,
});
exports.CredentialEntryStructList = koffi_1.default.struct('CredentialEntryStructList', {
    count: primitives_1.FFI_USIZE,
    data: koffi_1.default.pointer(exports.CredentialEntryStruct),
});
exports.CredentialProveStruct = koffi_1.default.struct('CredentialProveStruct', {
    entry_idx: primitives_1.FFI_INT64,
    referent: primitives_1.FFI_STRING,
    is_predicate: primitives_1.FFI_INT8,
    reveal: primitives_1.FFI_INT8,
});
exports.CredentialProveStructList = koffi_1.default.struct('CredentialProveStructList', {
    count: primitives_1.FFI_USIZE,
    data: koffi_1.default.pointer(exports.CredentialProveStruct),
});
exports.ObjectHandleStructList = koffi_1.default.struct('ObjectHandleStructList', {
    count: primitives_1.FFI_USIZE,
    data: koffi_1.default.pointer(primitives_1.FFI_OBJECT_HANDLE),
});
exports.NonrevokedIntervalOverrideStruct = koffi_1.default.struct('NonrevokedIntervalOverrideStruct', {
    rev_reg_def_id: primitives_1.FFI_STRING,
    requested_from_ts: primitives_1.FFI_INT32,
    override_rev_status_list_ts: primitives_1.FFI_INT32,
});
exports.NonrevokedIntervalOverrideStructList = koffi_1.default.struct('NonrevokedIntervalOverrideStructList', {
    count: primitives_1.FFI_USIZE,
    data: koffi_1.default.pointer(exports.NonrevokedIntervalOverrideStruct),
});
