// @credo-ts/core/build/modules/proofs/repository/ProofExchangeRecord.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProofExchangeRecord = void 0;
const error_1 = require("../../../error");
const BaseRecord_1 = require("../../../storage/BaseRecord");
const uuid_1 = require("../../../utils/uuid");
class ProofExchangeRecord extends BaseRecord_1.BaseRecord {
    constructor(props) {
        super();
        this.type = ProofExchangeRecord.type;
        if (props) {
            this.id = props.id ?? (0, uuid_1.uuid)();
            this.createdAt = props.createdAt ?? new Date();
            this.protocolVersion = props.protocolVersion;
            this.isVerified = props.isVerified;
            this.state = props.state;
            this.role = props.role;
            this.connectionId = props.connectionId;
            this.threadId = props.threadId;
            this.parentThreadId = props.parentThreadId;
            this.autoAcceptProof = props.autoAcceptProof;
            this._tags = props.tags ?? {};
            this.errorMessage = props.errorMessage;
        }
    }
    getTags() {
        return {
            ...this._tags,
            threadId: this.threadId,
            parentThreadId: this.parentThreadId,
            connectionId: this.connectionId,
            state: this.state,
            role: this.role,
        };
    }
    assertState(expectedStates) {
        if (!Array.isArray(expectedStates)) {
            expectedStates = [expectedStates];
        }
        if (!expectedStates.includes(this.state)) {
            throw new error_1.CredoError(`Proof record is in invalid state ${this.state}. Valid states are: ${expectedStates.join(', ')}.`);
        }
    }
    assertProtocolVersion(version) {
        if (this.protocolVersion !== version) {
            throw new error_1.CredoError(`Proof record has invalid protocol version ${this.protocolVersion}. Expected version ${version}`);
        }
    }
    assertConnection(currentConnectionId) {
        if (!this.connectionId) {
            throw new error_1.CredoError(`Proof record is not associated with any connection. This is often the case with connection-less presentation exchange`);
        }
        else if (this.connectionId !== currentConnectionId) {
            throw new error_1.CredoError(`Proof record is associated with connection '${this.connectionId}'. Current connection is '${currentConnectionId}'`);
        }
    }
}
exports.ProofExchangeRecord = ProofExchangeRecord;
ProofExchangeRecord.type = 'ProofRecord';
