// @hyperledger/anoncreds-nodejs/build/library/bindings.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.nativeBindings = void 0;
const koffi = __importStar(require("koffi"));
const ffi_1 = require("../ffi");
exports.nativeBindings = {
    // Version and utilities
    anoncreds_version: [koffi.pointer(ffi_1.FFI_STRING), []],
    anoncreds_set_default_logger: [ffi_1.FFI_ERROR_CODE, []],
    // Error handling
    anoncreds_get_current_error: [ffi_1.FFI_ERROR_CODE, [koffi.out(koffi.pointer(ffi_1.FFI_STRING))]],
    // Nonce
    anoncreds_generate_nonce: [ffi_1.FFI_ERROR_CODE, [koffi.out(koffi.pointer(ffi_1.FFI_STRING))]],
    // Buffer and memory management
    anoncreds_buffer_free: [ffi_1.FFI_VOID, [ffi_1.ByteBufferStruct]],
    anoncreds_string_free: [ffi_1.FFI_VOID, [koffi.pointer(ffi_1.FFI_STRING)]],
    anoncreds_object_free: [ffi_1.FFI_VOID, [ffi_1.FFI_OBJECT_HANDLE]],
    // Object operations
    anoncreds_object_get_json: [ffi_1.FFI_ERROR_CODE, [ffi_1.FFI_OBJECT_HANDLE, koffi.out(koffi.pointer(ffi_1.ByteBufferStruct))]],
    anoncreds_object_get_type_name: [ffi_1.FFI_ERROR_CODE, [ffi_1.FFI_OBJECT_HANDLE, koffi.out(koffi.pointer(ffi_1.FFI_STRING))]],
    // Schema operations
    anoncreds_create_schema: [
        ffi_1.FFI_ERROR_CODE,
        [ffi_1.FFI_STRING, ffi_1.FFI_STRING, ffi_1.FFI_STRING, ffi_1.StringListStruct, koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE))],
    ],
    anoncreds_schema_from_json: [ffi_1.FFI_ERROR_CODE, [ffi_1.ByteBufferStruct, koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE))]],
    // Credential definition operations
    anoncreds_create_credential_definition: [
        ffi_1.FFI_ERROR_CODE,
        [
            ffi_1.FFI_STRING,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_STRING,
            ffi_1.FFI_STRING,
            ffi_1.FFI_STRING,
            ffi_1.FFI_INT8,
            koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE)),
            koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE)),
            koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE)),
        ],
    ],
    anoncreds_credential_definition_from_json: [
        ffi_1.FFI_ERROR_CODE,
        [ffi_1.ByteBufferStruct, koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE))],
    ],
    anoncreds_credential_definition_private_from_json: [
        ffi_1.FFI_ERROR_CODE,
        [ffi_1.ByteBufferStruct, koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE))],
    ],
    // Key correctness proof
    anoncreds_key_correctness_proof_from_json: [
        ffi_1.FFI_ERROR_CODE,
        [ffi_1.ByteBufferStruct, koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE))],
    ],
    // Revocation registry operations
    anoncreds_create_revocation_registry_def: [
        ffi_1.FFI_ERROR_CODE,
        [
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_STRING,
            ffi_1.FFI_STRING,
            ffi_1.FFI_STRING,
            ffi_1.FFI_STRING,
            ffi_1.FFI_INT64,
            ffi_1.FFI_STRING,
            koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE)),
            koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE)),
        ],
    ],
    anoncreds_revocation_registry_definition_from_json: [
        ffi_1.FFI_ERROR_CODE,
        [ffi_1.ByteBufferStruct, koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE))],
    ],
    anoncreds_revocation_registry_definition_private_from_json: [
        ffi_1.FFI_ERROR_CODE,
        [ffi_1.ByteBufferStruct, koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE))],
    ],
    anoncreds_revocation_registry_definition_get_attribute: [
        ffi_1.FFI_ERROR_CODE,
        [ffi_1.FFI_OBJECT_HANDLE, ffi_1.FFI_STRING, koffi.out(koffi.pointer(ffi_1.FFI_STRING))],
    ],
    anoncreds_revocation_registry_from_json: [
        ffi_1.FFI_ERROR_CODE,
        [ffi_1.ByteBufferStruct, koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE))],
    ],
    // Revocation status list operations
    anoncreds_create_revocation_status_list: [
        ffi_1.FFI_ERROR_CODE,
        [
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_STRING,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_STRING,
            ffi_1.FFI_INT8,
            ffi_1.FFI_INT64,
            koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE)),
        ],
    ],
    anoncreds_update_revocation_status_list: [
        ffi_1.FFI_ERROR_CODE,
        [
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.I32ListStruct,
            ffi_1.I32ListStruct,
            ffi_1.FFI_INT64,
            koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE)),
        ],
    ],
    anoncreds_update_revocation_status_list_timestamp_only: [
        ffi_1.FFI_ERROR_CODE,
        [ffi_1.FFI_INT64, ffi_1.FFI_OBJECT_HANDLE, koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE))],
    ],
    anoncreds_revocation_status_list_from_json: [
        ffi_1.FFI_ERROR_CODE,
        [ffi_1.ByteBufferStruct, koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE))],
    ],
    // Revocation state operations
    anoncreds_create_or_update_revocation_state: [
        ffi_1.FFI_ERROR_CODE,
        [
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_INT64,
            ffi_1.FFI_STRING,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE)),
        ],
    ],
    anoncreds_revocation_state_from_json: [
        ffi_1.FFI_ERROR_CODE,
        [ffi_1.ByteBufferStruct, koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE))],
    ],
    // Link secret operations
    anoncreds_create_link_secret: [ffi_1.FFI_ERROR_CODE, [koffi.out(koffi.pointer(ffi_1.FFI_STRING))]],
    // Credential offer operations
    anoncreds_create_credential_offer: [
        ffi_1.FFI_ERROR_CODE,
        [ffi_1.FFI_STRING, ffi_1.FFI_STRING, ffi_1.FFI_OBJECT_HANDLE, koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE))],
    ],
    anoncreds_credential_offer_from_json: [
        ffi_1.FFI_ERROR_CODE,
        [ffi_1.ByteBufferStruct, koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE))],
    ],
    // Credential request operations
    anoncreds_create_credential_request: [
        ffi_1.FFI_ERROR_CODE,
        [
            ffi_1.FFI_STRING,
            ffi_1.FFI_STRING,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_STRING,
            ffi_1.FFI_STRING,
            ffi_1.FFI_OBJECT_HANDLE,
            koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE)),
            koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE)),
        ],
    ],
    anoncreds_credential_request_from_json: [
        ffi_1.FFI_ERROR_CODE,
        [ffi_1.ByteBufferStruct, koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE))],
    ],
    anoncreds_credential_request_metadata_from_json: [
        ffi_1.FFI_ERROR_CODE,
        [ffi_1.ByteBufferStruct, koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE))],
    ],
    // Credential operations
    anoncreds_create_credential: [
        ffi_1.FFI_ERROR_CODE,
        [
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.StringListStruct,
            ffi_1.StringListStruct,
            ffi_1.StringListStruct,
            koffi.pointer(ffi_1.CredRevInfoStruct),
            koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE)),
        ],
    ],
    anoncreds_credential_from_json: [ffi_1.FFI_ERROR_CODE, [ffi_1.ByteBufferStruct, koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE))]],
    anoncreds_credential_get_attribute: [
        ffi_1.FFI_ERROR_CODE,
        [ffi_1.FFI_OBJECT_HANDLE, ffi_1.FFI_STRING, koffi.out(koffi.pointer(ffi_1.FFI_STRING))],
    ],
    anoncreds_process_credential: [
        ffi_1.FFI_ERROR_CODE,
        [
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_STRING,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE)),
        ],
    ],
    anoncreds_encode_credential_attributes: [ffi_1.FFI_ERROR_CODE, [ffi_1.StringListStruct, koffi.out(koffi.pointer(ffi_1.FFI_STRING))]],
    // W3C Credential operations
    anoncreds_create_w3c_credential: [
        ffi_1.FFI_ERROR_CODE,
        [
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.StringListStruct,
            ffi_1.StringListStruct,
            koffi.pointer(ffi_1.CredRevInfoStruct),
            ffi_1.FFI_STRING,
            koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE)),
        ],
    ],
    anoncreds_w3c_credential_from_json: [ffi_1.FFI_ERROR_CODE, [ffi_1.ByteBufferStruct, koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE))]],
    anoncreds_credential_to_w3c: [
        ffi_1.FFI_ERROR_CODE,
        [ffi_1.FFI_OBJECT_HANDLE, ffi_1.FFI_STRING, ffi_1.FFI_STRING, koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE))],
    ],
    anoncreds_credential_from_w3c: [ffi_1.FFI_ERROR_CODE, [ffi_1.FFI_OBJECT_HANDLE, koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE))]],
    anoncreds_process_w3c_credential: [
        ffi_1.FFI_ERROR_CODE,
        [
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_STRING,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE)),
        ],
    ],
    anoncreds_w3c_credential_get_integrity_proof_details: [
        ffi_1.FFI_ERROR_CODE,
        [ffi_1.FFI_OBJECT_HANDLE, koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE))],
    ],
    anoncreds_w3c_credential_proof_get_attribute: [
        ffi_1.FFI_ERROR_CODE,
        [ffi_1.FFI_OBJECT_HANDLE, ffi_1.FFI_STRING, koffi.out(koffi.pointer(ffi_1.FFI_STRING))],
    ],
    // Presentation request operations
    anoncreds_presentation_request_from_json: [
        ffi_1.FFI_ERROR_CODE,
        [ffi_1.ByteBufferStruct, koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE))],
    ],
    // Presentation operations
    anoncreds_create_presentation: [
        ffi_1.FFI_ERROR_CODE,
        [
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.CredentialEntryStructList,
            ffi_1.CredentialProveStructList,
            ffi_1.StringListStruct,
            ffi_1.StringListStruct,
            ffi_1.FFI_STRING,
            ffi_1.ObjectHandleStructList,
            ffi_1.StringListStruct,
            ffi_1.ObjectHandleStructList,
            ffi_1.StringListStruct,
            koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE)),
        ],
    ],
    anoncreds_presentation_from_json: [ffi_1.FFI_ERROR_CODE, [ffi_1.ByteBufferStruct, koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE))]],
    anoncreds_verify_presentation: [
        ffi_1.FFI_ERROR_CODE,
        [
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.ObjectHandleStructList,
            ffi_1.StringListStruct,
            ffi_1.ObjectHandleStructList,
            ffi_1.StringListStruct,
            ffi_1.ObjectHandleStructList,
            ffi_1.StringListStruct,
            ffi_1.ObjectHandleStructList,
            ffi_1.NonrevokedIntervalOverrideStructList,
            koffi.out(koffi.pointer(ffi_1.FFI_INT8)),
        ],
    ],
    // W3C Presentation operations
    anoncreds_create_w3c_presentation: [
        ffi_1.FFI_ERROR_CODE,
        [
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.CredentialEntryStructList,
            ffi_1.CredentialProveStructList,
            ffi_1.FFI_STRING,
            ffi_1.ObjectHandleStructList,
            ffi_1.StringListStruct,
            ffi_1.ObjectHandleStructList,
            ffi_1.StringListStruct,
            ffi_1.FFI_STRING,
            koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE)),
        ],
    ],
    anoncreds_w3c_presentation_from_json: [
        ffi_1.FFI_ERROR_CODE,
        [ffi_1.ByteBufferStruct, koffi.out(koffi.pointer(ffi_1.FFI_OBJECT_HANDLE))],
    ],
    anoncreds_verify_w3c_presentation: [
        ffi_1.FFI_ERROR_CODE,
        [
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.ObjectHandleStructList,
            ffi_1.StringListStruct,
            ffi_1.ObjectHandleStructList,
            ffi_1.StringListStruct,
            ffi_1.ObjectHandleStructList,
            ffi_1.StringListStruct,
            ffi_1.ObjectHandleStructList,
            ffi_1.NonrevokedIntervalOverrideStructList,
            koffi.out(koffi.pointer(ffi_1.FFI_INT8)),
        ],
    ],
};
