// @hyperledger/aries-askar-nodejs/build/ffi/alloc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deallocateCallbackBuffer = exports.allocateCallbackBuffer = exports.allocateStringListHandle = exports.allocateLocalKeyHandle = exports.allocateAeadParams = exports.allocateEncryptedBuffer = exports.allocateSecretBuffer = exports.allocateInt8Buffer = exports.allocateInt32Buffer = exports.allocateStringBuffer = exports.allocatePointer = void 0;
const ref_napi_1 = require("@2060.io/ref-napi");
const primitives_1 = require("./primitives");
const structures_1 = require("./structures");
const allocatePointer = () => (0, ref_napi_1.alloc)(primitives_1.FFI_POINTER);
exports.allocatePointer = allocatePointer;
const allocateStringBuffer = () => (0, ref_napi_1.alloc)(primitives_1.FFI_STRING);
exports.allocateStringBuffer = allocateStringBuffer;
const allocateInt32Buffer = () => (0, ref_napi_1.alloc)(primitives_1.FFI_INT32);
exports.allocateInt32Buffer = allocateInt32Buffer;
const allocateInt8Buffer = () => (0, ref_napi_1.alloc)(primitives_1.FFI_INT8);
exports.allocateInt8Buffer = allocateInt8Buffer;
const allocateSecretBuffer = () => (0, ref_napi_1.alloc)(structures_1.SecretBufferStruct);
exports.allocateSecretBuffer = allocateSecretBuffer;
const allocateEncryptedBuffer = () => (0, ref_napi_1.alloc)(structures_1.EncryptedBufferStruct);
exports.allocateEncryptedBuffer = allocateEncryptedBuffer;
const allocateAeadParams = () => (0, ref_napi_1.alloc)(structures_1.AeadParamsStruct);
exports.allocateAeadParams = allocateAeadParams;
exports.allocateLocalKeyHandle = exports.allocatePointer;
exports.allocateStringListHandle = exports.allocatePointer;
const allocateCallbackBuffer = (callback) => setTimeout(() => callback, 1000000);
exports.allocateCallbackBuffer = allocateCallbackBuffer;
const deallocateCallbackBuffer = (id) => clearTimeout(id);
exports.deallocateCallbackBuffer = deallocateCallbackBuffer;
