// @hyperledger/aries-askar-nodejs/build/ffi/conversion.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.encryptedBufferStructToClass = exports.secretBufferToBuffer = exports.byteBufferToBuffer = exports.uint8arrayToByteBufferStruct = exports.secretBufferClassToStruct = exports.byteBufferClassToStruct = void 0;
const ref_napi_1 = require("@2060.io/ref-napi");
const aries_askar_shared_1 = require("@hyperledger/aries-askar-shared");
const structures_1 = require("./structures");
const byteBufferClassToStruct = ({ len, data }) => {
    return (0, structures_1.ByteBufferStruct)({
        len,
        data: data,
    });
};
exports.byteBufferClassToStruct = byteBufferClassToStruct;
exports.secretBufferClassToStruct = exports.byteBufferClassToStruct;
const uint8arrayToByteBufferStruct = (buf) => {
    return (0, exports.byteBufferClassToStruct)({ data: buf, len: buf.length });
};
exports.uint8arrayToByteBufferStruct = uint8arrayToByteBufferStruct;
const byteBufferToBuffer = ({ data, len }) => (0, ref_napi_1.reinterpret)(data, len);
exports.byteBufferToBuffer = byteBufferToBuffer;
exports.secretBufferToBuffer = exports.byteBufferToBuffer;
const encryptedBufferStructToClass = ({ secretBuffer, tagPos, noncePos }) => {
    const buffer = Uint8Array.from((0, exports.secretBufferToBuffer)(secretBuffer));
    return new aries_askar_shared_1.EncryptedBuffer({ tagPos, noncePos, buffer });
};
exports.encryptedBufferStructToClass = encryptedBufferStructToClass;
