// @hyperledger/aries-askar-nodejs/build/ffi/structures.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AeadParamsStructPtr = exports.AeadParamsStruct = exports.EncryptedBufferStructPtr = exports.EncryptedBufferStruct = exports.SecretBufferStructPtr = exports.SecretBufferStruct = exports.ByteBufferStruct = exports.ByteBufferArrayPtr = exports.ByteBufferArray = void 0;
const ref_napi_1 = __importStar(require("@2060.io/ref-napi"));
const ref_array_di_1 = __importDefault(require("ref-array-di"));
const ref_struct_di_1 = __importDefault(require("ref-struct-di"));
const primitives_1 = require("./primitives");
const CStruct = (0, ref_struct_di_1.default)(ref_napi_1.default);
const CArray = (0, ref_array_di_1.default)(ref_napi_1.default);
exports.ByteBufferArray = CArray(primitives_1.FFI_UINT8);
exports.ByteBufferArrayPtr = (0, ref_napi_1.refType)(exports.ByteBufferArray);
exports.ByteBufferStruct = CStruct({
    len: primitives_1.FFI_INT64,
    data: exports.ByteBufferArrayPtr,
});
const ByteBufferStructPtr = ref_napi_1.default.refType(exports.ByteBufferStruct);
exports.SecretBufferStruct = exports.ByteBufferStruct;
exports.SecretBufferStructPtr = ByteBufferStructPtr;
exports.EncryptedBufferStruct = CStruct({
    secretBuffer: exports.SecretBufferStruct,
    tagPos: primitives_1.FFI_INT64,
    noncePos: primitives_1.FFI_INT64,
});
exports.EncryptedBufferStructPtr = ref_napi_1.default.refType(exports.EncryptedBufferStruct);
exports.AeadParamsStruct = CStruct({
    nonceLength: primitives_1.FFI_INT32,
    tagLength: primitives_1.FFI_INT32,
});
exports.AeadParamsStructPtr = ref_napi_1.default.refType(exports.AeadParamsStruct);
